<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Generalized Reporter Score-Based Enrichment Analysis for Omics Data • ReporterScore</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Generalized Reporter Score-Based Enrichment Analysis for Omics Data">
<meta property="og:description" content="Inspired by the classic RSA, we developed the improved Generalized Reporter Score-based Analysis (GRSA) method, implemented in the R package ReporterScore, along with comprehensive visualization methods and pathway databases. GRSA is a threshold-free method that works well with all types of biomedical features, such as genes, chemical compounds, and microbial species. Importantly, the GRSA supports multi-group and longitudinal experimental designs, because of the included multi-group-compatible statistical methods.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ReporterScore</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/index.html">Vignettes</a>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="https://asa12138.github.io/pcutils" class="external-link">pcutils</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/pctax" class="external-link">pctax</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/ReporterScore" class="external-link">ReporterScore</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/MetaNet" class="external-link">MetaNet</a>
    </li>
    <li>
      <a href="https://iphylo.net" class="external-link">iPhylo</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/plot4fun" class="external-link">plot4fun</a>
    </li>
  </ul>
</li>
<li>
  <a href="https://asa12138.github.io/" class="external-link">Asa_web</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Asa12138/ReporterScore" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
<form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="reporterscore-">ReporterScore <img src="reference/figures/ReporterScore.png" align="right" width="120"><a class="anchor" aria-label="anchor" href="#reporterscore-"></a>
</h1></div>
<!-- badges: start -->

<p>Inspired by the classic ‘RSA’, we developed the improved ‘Generalized Reporter Score-based Analysis (GRSA)’ method, implemented in the R package ‘ReporterScore’, along with comprehensive visualization methods and pathway databases.</p>
<p>‘GRSA’ is a threshold-free method that works well with all types of biomedical features, such as genes, chemical compounds, and microbial species. Importantly, the ‘GRSA’ supports <strong>multi-group and longitudinal experimental designs</strong>, because of the included multi-group-compatible statistical methods.</p>
<p><img src="reference/figures/1-workflow.png" width="100%"></p>
<p>The HTML documentation of the latest version is available at <a href="https://asa12138.github.io/ReporterScore/" class="external-link">Github page</a>.</p>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>To cite ReporterScore in publications use:</p>
<p>C. Peng, Q. Chen, S. Tan, X. Shen, C. Jiang, Generalized Reporter Score-based Enrichment Analysis for Omics Data. <em>Briefings in Bioinformatics</em> (2024). <a href="https://doi.org/10.1093/bib/bbae116" class="external-link uri">https://doi.org/10.1093/bib/bbae116</a>.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the released version of <code>ReporterScore</code> from <a href="https://CRAN.R-project.org" class="external-link">CRAN</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ReporterScore"</span><span class="op">)</span></span></code></pre></div>
<p>You can install the development version of <code>ReporterScore</code> from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Asa12138/pcutils"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Asa12138/ReporterScore"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="section level3">
<h3 id="id_1-inputdata-feature-abundance-table-and-metadata">1. Inputdata (Feature abundance table and metadata)<a class="anchor" aria-label="anchor" href="#id_1-inputdata-feature-abundance-table-and-metadata"></a>
</h3>
<ul>
<li>For transcriptomic, scRNA-seq, and related gene-based omics data of a specific species, a complete gene abundance table can be used.</li>
<li>For metagenomic and metatranscriptomic data, which involve many different species, a KO abundance table can be used, generated using Blast, Diamond, or KEGG official mapper software to align the reads or contigs to the KEGG or the EggNOG database</li>
<li>For metabolomic data, an annotated compound abundance table can be used, but the standardization of compound IDs (e.g., convert compound IDs to C numbers in the KEGG database) is required.</li>
</ul>
<div class="section level4">
<h4 id="format-of-abundance-table">Format of abundance table:<a class="anchor" aria-label="anchor" href="#format-of-abundance-table"></a>
</h4>
<p>⚠️<strong>Importantly, the input abundance table should not be prefiltered to retain the background information, as the ‘GRSA’ is a threshold-free method.</strong></p>
<ul>
<li>The rownames are feature ids (e.g. “K00001” (KEGG K number) if feature=“ko”; “PEX11A” (gene symbol) if feature=“gene”; “C00024” (KEGG C number) if feature=“compound”).</li>
<li>The colnames are samples.</li>
<li>The abundance value can be read counts or normalized values (e.g., TPM, FPKM, RPKM, or relative abundance, corresponds to suitable statistical test method).</li>
</ul>
<p>An example code tailored for a KO abundance table is as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"KO_abundance_test"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">KO_abundance</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                WT1         WT2         WT3         WT4         WT5         WT6</span></span>
<span><span class="co">#&gt; K03169 0.002653545 0.005096380 0.002033923 0.000722349 0.003468322 0.001483028</span></span>
<span><span class="co">#&gt; K07133 0.000308237 0.000280458 0.000596527 0.000859854 0.000308719 0.000878098</span></span>
<span><span class="co">#&gt; K03088 0.002147068 0.002030742 0.003797459 0.004161979 0.002076596 0.003091182</span></span>
<span><span class="co">#&gt; K03530 0.003788366 0.000239298 0.000445817 0.000557271 0.000222969 0.000529624</span></span>
<span><span class="co">#&gt; K06147 0.000785654 0.001213630 0.001312569 0.001662740 0.002387006 0.001725797</span></span>
<span><span class="co">#&gt; K05349 0.001816325 0.002813642 0.003274701 0.001089906 0.002371921 0.001795214</span></span></code></pre></div>
<p>And you should also offer a experimental metadata:</p>
</div>
<div class="section level4">
<h4 id="format-of-metadata-table">Format of metadata table:<a class="anchor" aria-label="anchor" href="#format-of-metadata-table"></a>
</h4>
<ul>
<li>The rownames are samples, columns are groups.</li>
<li>The grouping variable can be categories (at least two categories, for differential abundance analysis).</li>
<li>The grouping variable can also be multiple time points (for longitudinal analysis).</li>
<li>The grouping variable can also be continuous (for correlation analysis).</li>
</ul>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Group Group2</span></span>
<span><span class="co">#&gt; WT1    WT     G3</span></span>
<span><span class="co">#&gt; WT2    WT     G3</span></span>
<span><span class="co">#&gt; WT3    WT     G3</span></span>
<span><span class="co">#&gt; WT4    WT     G3</span></span>
<span><span class="co">#&gt; WT5    WT     G3</span></span>
<span><span class="co">#&gt; WT6    WT     G1</span></span></code></pre></div>
<p>⚠️<strong>Importantly, the rownames of metadata and the colnames of feature abundance table should be matching or partial matching!</strong></p>
<p>The <code>ReporterScore</code> will automatically match the samples based on the rownames of metadata and the colnames of feature abundance table.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">KO_abundance</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">KO_abundance</span><span class="op">)</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span></span>
<span><span class="co">## TRUE</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="id_2-pathway-databases">2. Pathway databases<a class="anchor" aria-label="anchor" href="#id_2-pathway-databases"></a>
</h3>
<p>The <code>ReporterScore</code> package has built-in KEGG pathway, module, gene, compound, and GO databases and also allows customized databases, making it compatible with feature abundance tables from diverse omics data.</p>
<ol style="list-style-type: decimal">
<li><p><code>ReporterScore</code> has built-in KEGG pathway-KO and module-KO databases (2023-08 version) for KO abundance table. You can use <code><a href="reference/load_htable.html">load_KOlist()</a></code> to have a look and use <code><a href="reference/update_KEGG.html">update_KO_file()</a></code> to update these databases (by KEGG API) as using latest database is very important.</p></li>
<li><p><code>ReporterScore</code> has built-in KEGG pathway-compound and module-compound databases (2023-08 version) for compound abundance table. You can use <code><a href="reference/load_htable.html">load_CPDlist()</a></code> to have a look and use <code><a href="reference/update_KEGG.html">update_KO_file()</a></code> to update these databases (by KEGG API).</p></li>
<li><p><code>ReporterScore</code> has built-in pathway-ko, pathway-gene, and pathway-compound databases of human (hsa) and mouse (mmu) for ko/gene/compound abundance table. You can use <code><a href="reference/custom_modulelist_from_org.html">custom_modulelist_from_org()</a></code> to have a look. Use <code><a href="reference/update_KEGG.html">update_org_pathway()</a></code> to update these databases and download other organism databases (by KEGG API).</p></li>
<li><p><code>ReporterScore</code> has built-in GO-gene database, You can use <code><a href="reference/load_GOlist.html">load_GOlist()</a></code> to have a look and use <code><a href="reference/update_GOlist.html">update_GOlist()</a></code> to update these databases (by KEGG API).</p></li>
<li><p>You can just customize your own pathway databases (gene set of interest) by using <code><a href="reference/custom_modulelist.html">custom_modulelist()</a></code>.</p></li>
</ol>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. KEGG pathway-KO and module-KO databases</span></span>
<span><span class="va">KOlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_htable.html">load_KOlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">KOlist</span><span class="op">$</span><span class="va">pathway</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. KEGG pathway-compound and module-compound databases</span></span>
<span><span class="va">CPDlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_htable.html">load_CPDlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CPDlist</span><span class="op">$</span><span class="va">pathway</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. human (hsa) pathway-ko/gene/compound databases</span></span>
<span><span class="va">hsa_pathway_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/custom_modulelist_from_org.html">custom_modulelist_from_org</a></span><span class="op">(</span></span>
<span>  org <span class="op">=</span> <span class="st">"hsa"</span>,</span>
<span>  feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ko"</span>, <span class="st">"gene"</span>, <span class="st">"compound"</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa_pathway_gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. GO-gene database</span></span>
<span><span class="va">GOlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_GOlist.html">load_GOlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GOlist</span><span class="op">$</span><span class="va">BP</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 5. customize your own pathway databases</span></span>
<span><span class="op">?</span><span class="fu"><a href="reference/custom_modulelist.html">custom_modulelist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_3-one-step-grsa">3. One step GRSA<a class="anchor" aria-label="anchor" href="#id_3-one-step-grsa"></a>
</h3>
<p>Use function <code>GRSA</code> or <code>reporter_score</code> can get the reporter score result by one step.</p>
<p>⚠️There are some important arguments for analysis:</p>
<ul>
<li>
<strong>mode</strong>: “mixed” or “directed” (directed mode only for two groups differential analysis or multi-groups correlation analysis.), see details in <code>pvalue2zs</code>.</li>
<li>
<strong>method</strong>: the method of statistical test for calculating p-value. Default is <code>wilcox.test</code>:
<ul>
<li>
<code>t.test</code> (parametric) and <code>wilcox.test</code> (non-parametric). Perform comparison between two groups of samples.</li>
<li>
<code>anova</code> (parametric) and <code>kruskal.test</code> (non-parametric). Perform one-way ANOVA test or Kruskal-Wallis rank sum test comparing multiple groups.</li>
<li>“pearson”, “kendall”, or “spearman” (correlation), see <code><a href="https://rdrr.io/r/stats/cor.html" class="external-link">?cor</a></code>.</li>
<li>“none”: use “none” for <code>step by step enrichment</code>. You can calculate the p-value by other methods like “DESeq2”, “Edger”, “Limma”, “ALDEX”, “ANCOM” yourself.</li>
</ul>
</li>
<li>
<strong>type</strong>: choose the built-in pathway database:
<ul>
<li>‘pathway’ or ‘module’ for default KEGG database for <strong>microbiome</strong>.</li>
<li>‘CC’, ‘MF’, ‘BP’, ‘ALL’ for default GO database for <strong>homo sapiens</strong>.</li>
<li>org in listed in <a href="https://www.genome.jp/kegg/catalog/org_list.html" class="external-link uri">https://www.genome.jp/kegg/catalog/org_list.html</a> such as ‘hsa’ (if your kodf is come from a specific organism, you should specify type here)</li>
</ul>
</li>
<li>
<strong>modulelist</strong>: customize database. A dataframe containing ‘id’,‘K_num’,‘KOs’,‘Description’ columns. Take the <code>KOlist</code> as example, use <code>custom_modulelist</code> to build a customize database.</li>
<li>
<strong>feature</strong>: one of “ko”, “gene”, “compound”.</li>
</ul>
<p><strong>The first level will be set as the control group, you can change the factor level to change your comparison.</strong></p>
<p>For example, we want to compare two groups ‘WT-OE’, and use the “directed” mode as we just want know which pathways are enriched or depleted in <strong>OE group</strong>:</p>
<div class="section level4">
<h4 id="ko-pathway">KO-pathway<a class="anchor" aria-label="anchor" href="#ko-pathway"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Comparison: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Comparison:  WT OE</span></span>
<span></span>
<span><span class="co"># for microbiome!!!</span></span>
<span><span class="va">reporter_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"pathway"</span>, feature <span class="op">=</span> <span class="st">"ko"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ================================Use feature: ko=================================</span></span>
<span><span class="co">#&gt; ===============================Checking rownames================================</span></span>
<span><span class="co">#&gt; Some of your ko_stat are not KO id, check the format! (e.g. K00001)</span></span>
<span><span class="co">#&gt; 52.7% of your kos in the modulelist!</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ===================================1.KO test====================================</span></span>
<span><span class="co">#&gt; =================================Checking group=================================</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ==============================Calculating each KO===============================</span></span>
<span><span class="co">#&gt; ===========================Using method: wilcox.test============================</span></span>
<span><span class="co">#&gt; 1000 features done.</span></span>
<span><span class="co">#&gt; 2000 features done.</span></span>
<span><span class="co">#&gt; 3000 features done.</span></span>
<span><span class="co">#&gt; 4000 features done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Compared groups: WT, OE</span></span>
<span><span class="co">#&gt; Total KO number: 4535</span></span>
<span><span class="co">#&gt; Compare method: wilcox.test</span></span>
<span><span class="co">#&gt; Time use: 1.179</span></span>
<span><span class="co">#&gt; =========================2.Transfer p.value to Z-score==========================</span></span>
<span><span class="co">#&gt; ==========================3.Calculating reporter score==========================</span></span>
<span><span class="co">#&gt; ==================================load KOlist===================================</span></span>
<span><span class="co">#&gt; ===================KOlist download time: 2023-08-14 16:00:52====================</span></span>
<span><span class="co">#&gt; If you want to update KOlist, use `update_KO_file()`</span></span>
<span><span class="co">#&gt; ============================Calculating each pathway============================</span></span>
<span><span class="co">#&gt; 100 pathways done.</span></span>
<span><span class="co">#&gt; 400 pathways done.</span></span>
<span><span class="co">#&gt; ID number: 481</span></span>
<span><span class="co">#&gt; Time use: 1.689</span></span>
<span><span class="co">#&gt; ====================================All done====================================</span></span></code></pre></div>
<p>The result is a “reporter_score” object:</p>
<table class="table">
<thead><tr class="header">
<th>elements</th>
<th>description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>kodf</code></td>
<td>your input KO_abundance table</td>
</tr>
<tr class="even">
<td><code>ko_stat</code></td>
<td>ko statistics result contains p.value and z_score</td>
</tr>
<tr class="odd">
<td><code>reporter_s</code></td>
<td>the reporter score in each pathway</td>
</tr>
<tr class="even">
<td><code>modulelist</code></td>
<td>default KOlist or customized modulelist dataframe</td>
</tr>
<tr class="odd">
<td><code>group</code></td>
<td>The comparison groups in your data</td>
</tr>
<tr class="even">
<td><code>metadata</code></td>
<td>sample information dataframe contains group</td>
</tr>
</tbody>
</table>
</div>
<div class="section level4">
<h4 id="gene-pathway">Gene-pathway<a class="anchor" aria-label="anchor" href="#gene-pathway"></a>
</h4>
<p>When you use the gene abundance table of a specific species (e.g. human), remember to set the <code>feature</code> and <code>type</code>!!! Or give the database through <code>modulelist</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"genedf"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Method 1: Set the `feature` and `type`!</span></span>
<span><span class="va">reporter_res_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">genedf</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"hsa"</span>, feature <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ===============================Use feature: gene================================</span></span>
<span><span class="co">#&gt; ===============================Checking rownames================================</span></span>
<span><span class="co">#&gt; please make sure your input table rows are gene symbol!</span></span>
<span><span class="co">#&gt; 100% of your genes in the modulelist!</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ===================================1.KO test====================================</span></span>
<span><span class="co">#&gt; =================================Checking group=================================</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ==============================Calculating each KO===============================</span></span>
<span><span class="co">#&gt; ===========================Using method: wilcox.test============================</span></span>
<span><span class="co">#&gt; 1000 features done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Compared groups: WT, OE</span></span>
<span><span class="co">#&gt; Total KO number: 1000</span></span>
<span><span class="co">#&gt; Compare method: wilcox.test</span></span>
<span><span class="co">#&gt; Time use: 0.224</span></span>
<span><span class="co">#&gt; =========================2.Transfer p.value to Z-score==========================</span></span>
<span><span class="co">#&gt; ==========================3.Calculating reporter score==========================</span></span>
<span><span class="co">#&gt; ================================load hsa pathway================================</span></span>
<span><span class="co">#&gt; =================hsa pathway download time: 2023-08-14 23:28:13=================</span></span>
<span><span class="co">#&gt; If you want to update hsa pathway, use `update_org_pathway('hsa')`</span></span>
<span><span class="co">#&gt; please assgin this custom modulelist to `reporter_score(modulelist=your_modulelist)` to do a custom enrichment.</span></span>
<span><span class="co">#&gt; You choose the feature: 'gene', make sure the rownames of your input table are right.</span></span>
<span><span class="co">#&gt; ============================Calculating each pathway============================</span></span>
<span><span class="co">#&gt; 100 pathways done.</span></span>
<span><span class="co">#&gt; 150 pathways done.</span></span>
<span><span class="co">#&gt; 200 pathways done.</span></span>
<span><span class="co">#&gt; 250 pathways done.</span></span>
<span><span class="co">#&gt; 300 pathways done.</span></span>
<span><span class="co">#&gt; ID number: 343</span></span>
<span><span class="co">#&gt; Time use: 1.558</span></span>
<span><span class="co">#&gt; ====================================All done====================================</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Method 2: Give the database through `modulelist`, same to Method 1.</span></span>
<span><span class="va">hsa_pathway_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/custom_modulelist_from_org.html">custom_modulelist_from_org</a></span><span class="op">(</span></span>
<span>  org <span class="op">=</span> <span class="st">"hsa"</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">reporter_res_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">genedf</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  modulelist <span class="op">=</span> <span class="va">hsa_pathway_gene</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res_gene</span>, rs_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use `modify_description` to remove the suffix of pathway description</span></span>
<span><span class="va">reporter_res_gene2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/modify_description.html">modify_description</a></span><span class="op">(</span><span class="va">reporter_res_gene</span>, pattern <span class="op">=</span> <span class="st">" - Homo sapiens (human)"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res_gene2</span>, rs_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use `ggplot_translator` to translate pathway description</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">pcutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/ggplot_translator.html" class="external-link">ggplot_translator</a></span><span class="op">(</span><span class="va">p2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Please set the font family to make the labels display well.</span></span>
<span><span class="co">#&gt;  see `how_to_set_font_for_plot()`.</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span> <span class="op">/</span> <span class="va">p3</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-10-1.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="compound-pathway">Compound-pathway<a class="anchor" aria-label="anchor" href="#compound-pathway"></a>
</h4>
<p>When you use the compound abundance table, remember to set the <code>feature</code> and <code>type</code>!!! Or give the database through <code>modulelist</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reporter_res_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">chem_df</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"hsa"</span>, feature <span class="op">=</span> <span class="st">"compound"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="id_4-visualization">4. Visualization<a class="anchor" aria-label="anchor" href="#id_4-visualization"></a>
</h3>
<p>After we get the reporter score result, we can visualize the result in various ways.</p>
<p>When we focus on the whole result:</p>
<ul>
<li>Plot the most significantly enriched pathways:</li>
</ul>
<p>You can set the <code>rs_threshold</code> to filter the pathways, the default <code>rs_threshold</code> is 1.64, which is be considered as significant at the 0.05 level.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(reporter_res$reporter_s)</span></span>
<span><span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res</span>, rs_threshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span><span class="op">)</span>, facet_level <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; ==============================load Pathway_htable===============================</span></span>
<span><span class="co">#&gt; ===============Pathway_htable download time: 2024-01-12 00:52:39================</span></span>
<span><span class="co">#&gt; If you want to update Pathway_htable, use `update_htable(type='pathway')`</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-12-1.png" width="100%"></p>
<p>⚠️<strong>In the directed mode, Enriched in one group means depleted in another group.</strong></p>
<ul>
<li>Plot the most significantly enriched pathways (circle packing):</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_report_circle_packing.html">plot_report_circle_packing</a></span><span class="op">(</span><span class="va">reporter_res</span>, rs_threshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; ==============================load Pathway_htable===============================</span></span>
<span><span class="co">#&gt; ===============Pathway_htable download time: 2024-01-12 00:52:39================</span></span>
<span><span class="co">#&gt; If you want to update Pathway_htable, use `update_htable(type='pathway')`</span></span>
<span><span class="co">#&gt; Non-leaf weights ignored</span></span>
<span><span class="co">#&gt; Scale for fill is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for fill, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-13-1.png" width="100%"></p>
<p>When we focus on one pathway, e.g. “map00780”:</p>
<ul>
<li>Plot boxes and lines</li>
</ul>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_in_pathway.html">plot_KOs_in_pathway</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-14-1.png" width="100%"></p>
<ul>
<li>Plot the distribution of Z-scores</li>
</ul>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_distribution.html">plot_KOs_distribution</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-15-1.png" width="100%"></p>
<ul>
<li>Plot as a network:</li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_network.html">plot_KOs_network</a></span><span class="op">(</span><span class="va">reporter_res</span>,</span>
<span>  map_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"map00780"</span>, <span class="st">"map00785"</span>, <span class="st">"map00900"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">""</span>, mark_module <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-1.png" width="100%"></p>
<ul>
<li>Plot the KOs abundance in a pathway:</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_box.html">plot_KOs_box</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span>, only_sig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-17-1.png" width="100%"></p>
<ul>
<li>Plot the KOs abundance in a pathway (heatmap):</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_heatmap.html">plot_KOs_heatmap</a></span><span class="op">(</span><span class="va">reporter_res</span>,</span>
<span>  map_id <span class="op">=</span> <span class="st">"map00780"</span>, only_sig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cutree_rows <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-18-1.png" width="100%"></p>
<ul>
<li>Plot the KEGG pathway map:</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_KEGG_map.html">plot_KEGG_map</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span>, color_var <span class="op">=</span> <span class="st">"Z_score"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/ko00780.Z_score.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="example-for-multi-group-or-longitudinal">Example for multi-group or longitudinal<a class="anchor" aria-label="anchor" href="#example-for-multi-group-or-longitudinal"></a>
</h3>
<p>If our experimental design is more than two groups or longitudinal, we can choose multi-groups comparison (or correlation):</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Comparison: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Group2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Comparison:  G1 G2 G3</span></span>
<span></span>
<span><span class="va">reporter_res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group2"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"spearman"</span>, perm <span class="op">=</span> <span class="fl">999</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ================================Use feature: ko=================================</span></span>
<span><span class="co">#&gt; ===============================Checking rownames================================</span></span>
<span><span class="co">#&gt; Some of your ko_stat are not KO id, check the format! (e.g. K00001)</span></span>
<span><span class="co">#&gt; 52.7% of your kos in the modulelist!</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ===================================1.KO test====================================</span></span>
<span><span class="co">#&gt; =================================Checking group=================================</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ==============================Calculating each KO===============================</span></span>
<span><span class="co">#&gt; =============================Using method: spearman=============================</span></span>
<span><span class="co">#&gt; Using correlation analysis: spearman, the groups will be transform to numeric, note the factor feature of group.</span></span>
<span><span class="co">#&gt; 1000 features done.</span></span>
<span><span class="co">#&gt; 2000 features done.</span></span>
<span><span class="co">#&gt; 3000 features done.</span></span>
<span><span class="co">#&gt; 4000 features done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Compared groups: G1, G2, G3</span></span>
<span><span class="co">#&gt; Total KO number: 4535</span></span>
<span><span class="co">#&gt; Compare method: spearman</span></span>
<span><span class="co">#&gt; Time use: 0.549</span></span>
<span><span class="co">#&gt; =========================2.Transfer p.value to Z-score==========================</span></span>
<span><span class="co">#&gt; ==========================3.Calculating reporter score==========================</span></span>
<span><span class="co">#&gt; ==================================load KOlist===================================</span></span>
<span><span class="co">#&gt; ===================KOlist download time: 2023-08-14 16:00:52====================</span></span>
<span><span class="co">#&gt; If you want to update KOlist, use `update_KO_file()`</span></span>
<span><span class="co">#&gt; ============================Calculating each pathway============================</span></span>
<span><span class="co">#&gt; 100 pathways done.</span></span>
<span><span class="co">#&gt; 400 pathways done.</span></span>
<span><span class="co">#&gt; ID number: 481</span></span>
<span><span class="co">#&gt; Time use: 1.643</span></span>
<span><span class="co">#&gt; ====================================All done====================================</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_features_in_pathway.html">plot_KOs_in_pathway</a></span><span class="op">(</span><span class="va">reporter_res2</span>, map_id <span class="op">=</span> <span class="st">"map02060"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-21-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="example-for-specified-pattern">Example for specified pattern<a class="anchor" aria-label="anchor" href="#example-for-specified-pattern"></a>
</h3>
<p>For example, groups “G1”, “G2”, and “G3” can be set as 1, 10, and 100 if an exponentially increasing trend is expected.</p>
<p>We use 1,5,1 to found pathways with the down-up-down pattern</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reporter_res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group2"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pearson"</span>, pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G1"</span> <span class="op">=</span> <span class="fl">1</span>, <span class="st">"G2"</span> <span class="op">=</span> <span class="fl">5</span>, <span class="st">"G3"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res3</span>, rs_threshold <span class="op">=</span> <span class="fl">3</span>, show_ID <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-1.png" width="100%"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_in_pathway.html">plot_KOs_in_pathway</a></span><span class="op">(</span><span class="va">reporter_res3</span>, map_id <span class="op">=</span> <span class="st">"map00860"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-2.png" width="100%"></p>
<p>To explore potential patterns within the data, clustering methods, such as C-means clustering, can be used.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rsa_cm_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RSA_by_cm.html">RSA_by_cm</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group2"</span>, <span class="va">metadata</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  k_num <span class="op">=</span> <span class="fl">3</span>, perm <span class="op">=</span> <span class="fl">999</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># show the patterns</span></span>
<span><span class="fu"><a href="reference/RSA_by_cm.html">plot_c_means</a></span><span class="op">(</span><span class="va">rsa_cm_res</span>, filter_membership <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-23-1.png" width="100%"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">rsa_cm_res</span>, rs_threshold <span class="op">=</span> <span class="fl">2.5</span>, y_text_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-23-2.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a>
</h2>
<div class="section level3">
<h3 id="step-by-step">Step by step<a class="anchor" aria-label="anchor" href="#step-by-step"></a>
</h3>
<p>The one step function <code>reporter_score</code> consists of three parts：</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"KO_abundance_test"</span><span class="op">)</span></span>
<span><span class="va">ko_pvalue</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ko.test.html">ko.test</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>, method <span class="op">=</span> <span class="st">"wilcox.test"</span><span class="op">)</span></span>
<span><span class="va">ko_stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/pvalue2zs.html">pvalue2zs</a></span><span class="op">(</span><span class="va">ko_pvalue</span>, mode <span class="op">=</span> <span class="st">"directed"</span><span class="op">)</span></span>
<span><span class="va">reporter_s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_reporter_score.html">get_reporter_score</a></span><span class="op">(</span><span class="va">ko_stat</span>, perm <span class="op">=</span> <span class="fl">499</span><span class="op">)</span></span></code></pre></div>
<ol style="list-style-type: decimal">
<li>
<code>ko.test</code>: this function help to calculate <em>p-value</em> for KO_abundance by various built-in methods such as differential analysis (<code>t.test</code>, <code>wilcox.test</code>, <code>kruskal.test</code>, <code>anova</code>) or correlation analysis (<code>pearson</code>, <code>spearman</code>, <code>kendall</code>). <strong>You can also calculate this <em>p-value</em> for KO_abundance by other methods</strong> like “DESeq2”, “Edger”, “Limma”, “ALDEX”, “ANCOM” and do a p.adjust yourself, then skip <code>ko.test</code> step go to step2…</li>
<li>
<code>pvalue2zs</code>: this function transfers p-value of KOs to Z-score (select mode: “mixed” or “directed”).</li>
<li>
<code>get_reporter_score</code> this function calculate reporter score of each pathways in a specific database. You can use a custom database here.</li>
</ol>
<p>Take the “Limma” as an example:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1-1. Calculate p-value by Limma</span></span>
<span><span class="va">ko_pvalue</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ko.test.html">ko.test</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>, method <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ko_Limma_p</span> <span class="op">&lt;-</span> <span class="fu">pctax</span><span class="fu">::</span><span class="fu">diff_da</span><span class="op">(</span><span class="va">KO_abundance</span>, group_df <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="st">"Group"</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"limma"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1-2. Replace the p-value in ko_pvalue, remember to match the KO_ids</span></span>
<span><span class="va">ko_pvalue</span><span class="op">$</span><span class="va">`p.value`</span> <span class="op">&lt;-</span> <span class="va">ko_Limma_p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ko_pvalue</span><span class="op">$</span><span class="va">KO_id</span>, <span class="va">ko_Limma_p</span><span class="op">$</span><span class="va">tax</span><span class="op">)</span>, <span class="st">"pvalue"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. Use `pvalue2zs` to get Z-score</span></span>
<span><span class="va">ko_stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/pvalue2zs.html">pvalue2zs</a></span><span class="op">(</span><span class="va">ko_pvalue</span>, mode <span class="op">=</span> <span class="st">"directed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. Use `get_reporter_score` to get reporter score</span></span>
<span><span class="va">reporter_s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_reporter_score.html">get_reporter_score</a></span><span class="op">(</span><span class="va">ko_stat</span>, perm <span class="op">=</span> <span class="fl">499</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. Combine the result</span></span>
<span><span class="va">reporter_res1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/combine_rs_res.html">combine_rs_res</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>, <span class="va">ko_stat</span>, <span class="va">reporter_s1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Then the reporter_res1 can be used for visualization</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="other-commonly-used-enrichment-methods">Other commonly used enrichment methods<a class="anchor" aria-label="anchor" href="#other-commonly-used-enrichment-methods"></a>
</h3>
<table class="table">
<colgroup>
<col width="3%">
<col width="23%">
<col width="17%">
<col width="55%">
</colgroup>
<thead><tr class="header">
<th align="left">Category</th>
<th align="left">Method</th>
<th align="left">Tools</th>
<th align="left">Notes</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">ORA</td>
<td align="left">Hypergeometric test / Fisher’s exact test</td>
<td align="left">DAVID (website) , clusterProfiler (R package)</td>
<td align="left">The most common methods used in enrichment analysis. Selecting a list of genes is required.</td>
</tr>
<tr class="even">
<td align="left">FCS</td>
<td align="left">Gene set enrichment analysis (GSEA)</td>
<td align="left">GSEA (website)</td>
<td align="left">GSEA creatively uses gene ranking, rather than selecting a list of genes, to identify statistically significant and concordant differences across gene sets.</td>
</tr>
<tr class="odd">
<td align="left">FCS</td>
<td align="left">Generalized reporter score-based analysis (GRSA/RSA)</td>
<td align="left">ReporterScore (R package developed in this study)</td>
<td align="left">Find significant metabolites (first report), pathways, and taxonomy based on the p-values for multi-omics data.</td>
</tr>
<tr class="even">
<td align="left">FCS</td>
<td align="left">Significance Analysis of Function and Expression (SAFE)</td>
<td align="left">safe (R package)</td>
<td align="left">SAFE assesses the significance of gene categories by calculating both local and global statistics from gene expression data.</td>
</tr>
<tr class="odd">
<td align="left">FCS</td>
<td align="left">Gene Set Analysis (GSA)</td>
<td align="left">GSA (R Package)</td>
<td align="left">GSA was proposed as an improvement of GSEA, using the “maxmean” statistic instead of the weighted sign KS statistic.</td>
</tr>
<tr class="even">
<td align="left">FCS</td>
<td align="left">Pathway Analysis with Down-weighting of Overlapping Genes (PADOG)</td>
<td align="left">PADOG (R package)</td>
<td align="left">PADOGA assumes that genes associated with fewer pathways have more significant effects than genes associated with more pathways.</td>
</tr>
<tr class="odd">
<td align="left">FCS</td>
<td align="left">Gene Set Variation Analysis (GSVA)</td>
<td align="left">GSVA (R package)</td>
<td align="left">A nonparametric, unsupervised method that transforms gene expression data into gene set scores for downstream differential pathway activity analysis.</td>
</tr>
<tr class="even">
<td align="left">PT</td>
<td align="left">Topology-based pathway enrichment analysis (TPEA)</td>
<td align="left">TPEA (R package)</td>
<td align="left">Integrate topological properties and global upstream/downstream positions of genes in pathways.</td>
</tr>
</tbody>
</table>
<p>Commonly used enrichment methods for omics data.</p>
<p><code>ReporterScore</code> also provides other enrichment methods like <code>KO_fisher</code>(fisher.test), <code>KO_enrich</code>(fisher.test, from <code>clusterProfiler</code>), <code>KO_gsea</code> (GSEA, from <code>clusterProfiler</code>).</p>
<p>The input data is from <code>reporter_score</code>, and also supports custom databases, so you can easily compare the results of various enrichment methods and conduct a comprehensive analysis:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(reporter_res2$reporter_s)</span></span>
<span><span class="co"># reporter_score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">$</span><span class="va">reporter_s</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">ReporterScore</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1.64</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">RS</span></span>
<span><span class="co"># fisher</span></span>
<span><span class="va">fisher_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/KO_fisher.html">KO_fisher</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fisher_res</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">Fisher</span></span>
<span><span class="co"># enricher</span></span>
<span><span class="va">enrich_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/KO_enrich.html">KO_enrich</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">enrich_res</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">clusterProfiler</span></span>
<span><span class="co"># GESA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">gsea_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/KO_gsea.html">KO_gsea</a></span><span class="op">(</span><span class="va">reporter_res</span>, weight <span class="op">=</span> <span class="st">"Z_score"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, : There are ties in the preranked stats (57.15% of the list).</span></span>
<span><span class="co">#&gt; The order of those tied genes will be arbitrary, which may produce unexpected results.</span></span>
<span><span class="co">#&gt; Warning in fgseaMultilevel(pathways = pathways, stats = stats, minSize =</span></span>
<span><span class="co">#&gt; minSize, : For some pathways, in reality P-values are less than 1e-10. You can</span></span>
<span><span class="co">#&gt; set the `eps` argument to zero for better estimation.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">gsea_res</span><span class="op">)</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">GSEA</span></span>
<span></span>
<span><span class="va">venn_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>GRSA <span class="op">=</span> <span class="va">RS</span>, Fisher <span class="op">=</span> <span class="va">Fisher</span>, CP <span class="op">=</span> <span class="va">clusterProfiler</span>, GSEA <span class="op">=</span> <span class="va">GSEA</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Asa12138/pcutils" class="external-link">pcutils</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/venn.html" class="external-link">venn</a></span><span class="op">(</span><span class="va">venn_res</span>, <span class="st">"network"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-25-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="other-features">Other features<a class="anchor" aria-label="anchor" href="#other-features"></a>
</h2>
<div class="section level3">
<h3 id="uplevel-the-kos">uplevel the KOs<a class="anchor" aria-label="anchor" href="#uplevel-the-kos"></a>
</h3>
<p><a href="https://www.genome.jp/kegg/brite.html" class="external-link">KEGG BRITE</a> is a collection of hierarchical classification systems capturing functional hierarchies of various biological objects, especially those represented as KEGG objects.</p>
<p>We collected k00001 KEGG Orthology (KO) table so that you can summaries each levels abundance. Use <code>load_KO_htable</code> to get KO_htable and use <code>update_KO_htable</code> to update. Use <code>up_level_KO</code> can upgrade to specific level in one of “pathway”, “module”, “level1”, “level2”, “level3”, “module1”, “module2”, “module3”.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">KO_htable</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_htable.html">load_KO_htable</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load KO_htable=================================</span></span>
<span><span class="co">#&gt; ==================KO_htable download time: 2024-01-12 00:49:03==================</span></span>
<span><span class="co">#&gt; If you want to update KO_htable, use `update_htable(type='ko')`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">KO_htable</span><span class="op">)</span></span>
<span><span class="co">#&gt;   level1_name             level2_name level3_id                  level3_name</span></span>
<span><span class="co">#&gt; 1  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 2  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 3  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 4  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 5  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 6  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt;    KO_id                                                    KO_name</span></span>
<span><span class="co">#&gt; 1 K00844                                HK; hexokinase [EC:2.7.1.1]</span></span>
<span><span class="co">#&gt; 2 K12407                              GCK; glucokinase [EC:2.7.1.2]</span></span>
<span><span class="co">#&gt; 3 K00845                              glk; glucokinase [EC:2.7.1.2]</span></span>
<span><span class="co">#&gt; 4 K25026                              glk; glucokinase [EC:2.7.1.2]</span></span>
<span><span class="co">#&gt; 5 K01810       GPI, pgi; glucose-6-phosphate isomerase [EC:5.3.1.9]</span></span>
<span><span class="co">#&gt; 6 K06859 pgi1; glucose-6-phosphate isomerase, archaeal [EC:5.3.1.9]</span></span>
<span><span class="fu"><a href="reference/plot_htable.html">plot_htable</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"ko"</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load KO_htable=================================</span></span>
<span><span class="co">#&gt; ==================KO_htable download time: 2024-01-12 00:49:03==================</span></span>
<span><span class="co">#&gt; If you want to update KO_htable, use `update_htable(type='ko')`</span></span>
<span><span class="co">#&gt; Warning: Vectorized input to `element_text()` is not officially supported.</span></span>
<span><span class="co">#&gt; ℹ Results may be unexpected or may change in future versions of ggplot2.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-26-1.png" width="100%"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">KO_level1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/up_level_KO.html">up_level_KO</a></span><span class="op">(</span><span class="va">KO_abundance</span>, level <span class="op">=</span> <span class="st">"level1"</span>, show_name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load KO_htable=================================</span></span>
<span><span class="co">#&gt; ==================KO_htable download time: 2024-01-12 00:49:03==================</span></span>
<span><span class="co">#&gt; If you want to update KO_htable, use `update_htable(type='ko')`</span></span>
<span><span class="fu">pcutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/stackplot.html" class="external-link">stackplot</a></span><span class="op">(</span><span class="va">KO_level1</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">KO_level1</span><span class="op">)</span> <span class="op">==</span> <span class="st">"Unknown"</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggsci</span><span class="fu">::</span><span class="fu">scale_fill_d3</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-27-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="card-for-args">CARD for ARGs<a class="anchor" aria-label="anchor" href="#card-for-args"></a>
</h3>
<p>For convenience, I also included the CARD database from <a href="https://card.mcmaster.ca/download/0/broadstreet-v3.2.8.tar.bz2" class="external-link uri">https://card.mcmaster.ca/download/0/broadstreet-v3.2.8.tar.bz2</a>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CARDinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_CARDinfo.html">load_CARDinfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load CARDinfo==================================</span></span>
<span><span class="co">#&gt; ==================CARDinfo download time: 2024-01-12 01:12:11===================</span></span>
<span><span class="co">#&gt; If you want to update CARDinfo, use `update_GOlist()`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CARDinfo</span><span class="op">$</span><span class="va">ARO_index</span><span class="op">)</span></span>
<span><span class="co">#&gt;         ARO Accession CVTERM ID Model Sequence ID Model ID</span></span>
<span><span class="co">#&gt; 3005099   ARO:3005099     43314              6143     3831</span></span>
<span><span class="co">#&gt; 3002523   ARO:3002523     38923              8144     1781</span></span>
<span><span class="co">#&gt; 3002524   ARO:3002524     38924                85      746</span></span>
<span><span class="co">#&gt; 3002525   ARO:3002525     38925              4719     1246</span></span>
<span><span class="co">#&gt; 3002526   ARO:3002526     38926               228     1415</span></span>
<span><span class="co">#&gt; 3002527   ARO:3002527     38927              5510     2832</span></span>
<span><span class="co">#&gt;                                                     Model Name</span></span>
<span><span class="co">#&gt; 3005099 23S rRNA (adenine(2058)-N(6))-methyltransferase Erm(A)</span></span>
<span><span class="co">#&gt; 3002523                                             AAC(2')-Ia</span></span>
<span><span class="co">#&gt; 3002524                                             AAC(2')-Ib</span></span>
<span><span class="co">#&gt; 3002525                                             AAC(2')-Ic</span></span>
<span><span class="co">#&gt; 3002526                                             AAC(2')-Id</span></span>
<span><span class="co">#&gt; 3002527                                             AAC(2')-Ie</span></span>
<span><span class="co">#&gt;                                                       ARO Name</span></span>
<span><span class="co">#&gt; 3005099 23S rRNA (adenine(2058)-N(6))-methyltransferase Erm(A)</span></span>
<span><span class="co">#&gt; 3002523                                             AAC(2')-Ia</span></span>
<span><span class="co">#&gt; 3002524                                             AAC(2')-Ib</span></span>
<span><span class="co">#&gt; 3002525                                             AAC(2')-Ic</span></span>
<span><span class="co">#&gt; 3002526                                             AAC(2')-Id</span></span>
<span><span class="co">#&gt; 3002527                                             AAC(2')-Ie</span></span>
<span><span class="co">#&gt;         Protein Accession DNA Accession                         AMR Gene Family</span></span>
<span><span class="co">#&gt; 3005099        AAB60941.1    AF002716.1 Erm 23S ribosomal RNA methyltransferase</span></span>
<span><span class="co">#&gt; 3002523        AAA03550.1      L06156.2                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002524        AAC44793.1      U41471.1                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002525        CCP42991.1    AL123456.3                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002526        AAB41701.1      U72743.1                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002527        CAC32082.1    AL583926.1                                 AAC(2')</span></span>
<span><span class="co">#&gt;                                                                   Drug Class</span></span>
<span><span class="co">#&gt; 3005099 lincosamide antibiotic;macrolide antibiotic;streptogramin antibiotic</span></span>
<span><span class="co">#&gt; 3002523                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002524                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002525                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002526                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002527                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt;                 Resistance Mechanism CARD Short Name</span></span>
<span><span class="co">#&gt; 3005099 antibiotic target alteration  Spyo_ErmA_MLSb</span></span>
<span><span class="co">#&gt; 3002523      antibiotic inactivation      AAC(2')-Ia</span></span>
<span><span class="co">#&gt; 3002524      antibiotic inactivation      AAC(2')-Ib</span></span>
<span><span class="co">#&gt; 3002525      antibiotic inactivation      AAC(2')-Ic</span></span>
<span><span class="co">#&gt; 3002526      antibiotic inactivation      AAC(2')-Id</span></span>
<span><span class="co">#&gt; 3002527      antibiotic inactivation      AAC(2')-Ie</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="reference">Reference<a class="anchor" aria-label="anchor" href="#reference"></a>
</h1>
<ol style="list-style-type: decimal">
<li><p>Patil, K. R. &amp; Nielsen, J. Uncovering transcriptional regulation of metabolism by using metabolic network topology. Proc Natl Acad Sci U S A 102, 2685–2689 (2005).</p></li>
<li><p>L. Liu, R. Zhu, D. Wu, Misuse of reporter score in microbial enrichment analysis. iMeta. 2, e95 (2023).</p></li>
<li><p><a href="https://github.com/wangpeng407/ReporterScore" class="external-link uri">https://github.com/wangpeng407/ReporterScore</a></p></li>
</ol>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=ReporterScore" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/Asa12138/ReporterScore/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/Asa12138/ReporterScore/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing ReporterScore</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Chen Peng <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-9449-7606" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/Asa12138/ReporterScore/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/Asa12138/ReporterScore/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://doi.org/10.1093/bib/bbae116" class="external-link"><img src="https://img.shields.io/badge/doi-10.1093/bib/bbae116-yellow.svg"></a></li>
<li><a href="https://asa-blog.netlify.app/" class="external-link"><img src="https://img.shields.io/badge/blog-@asa-blue.svg"></a></li>
<li><a href="https://cran.r-project.org/package=ReporterScore" class="external-link"><img src="http://cranlogs.r-pkg.org/badges/grand-total/ReporterScore"></a></li>
<li><a href="https://cran.r-project.org/package=ReporterScore" class="external-link"><img src="http://cranlogs.r-pkg.org/badges/last-month/ReporterScore"></a></li>
<li><a href="https://cran.r-project.org/package=ReporterScore" class="external-link"><img src="https://www.r-pkg.org/badges/version/ReporterScore?color=green"></a></li>
<li><a href="https://github.com/Asa12138/ReporterScore" class="external-link"><img src="https://img.shields.io/badge/devel%20version-0.1.5-green.svg"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Chen Peng.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: '599cec31baffa4868cae4e79f180729b',
    indexName: 'docsearch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
