<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ReporterScore  • ReporterScore</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png"><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css"><script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet"><script src="pkgdown.js"></script><!-- docsearch --><script src="docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.3/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous"><link href="docsearch.css" rel="stylesheet"><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="ReporterScore "><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-title-body">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">ReporterScore</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>

  </a>
</li>
<li>
  <a href="articles/index.html">Vignettes</a>
</li>
<li>
  <a href="reference/index.html">Functions</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other Packages

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="https://asa12138.github.io/pcutils" class="external-link">pcutils</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/pctax" class="external-link">pctax</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/ReporterScore" class="external-link">ReporterScore</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/MetaNet" class="external-link">MetaNet</a>
    </li>
    <li>
      <a href="https://iphylo.net" class="external-link">iPhylo</a>
    </li>
    <li>
      <a href="https://asa12138.github.io/plot4fun" class="external-link">plot4fun</a>
    </li>
  </ul></li>
<li>
  <a href="https://asa12138.github.io/" class="external-link">Asa_web</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/Asa12138/ReporterScore" class="external-link">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul><form class="navbar-form navbar-right hidden-xs hidden-sm" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off"></div>
      </form>

    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>ReporterScore </h1>
    </div>


<p><strong>其他语言版本: <a href="README.html">English</a>, <a href="README_zh_CN.html">中文</a>.</strong></p>
<div id="reporterscore-" class="section level1">

<!-- badges: start -->
<p><a href="https://github.com/Asa12138/ReporterScore/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/Asa12138/ReporterScore/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a> <a href="https://doi.org/10.1093/bib/bbae116" class="external-link"><img src="https://img.shields.io/badge/doi-10.1093/bib/bbae116-yellow.svg"></a> <a href="https://asa-blog.netlify.app/" class="external-link"><img src="https://img.shields.io/badge/blog-@asa-blue.svg"></a> <a href="https://cran.r-project.org/package=ReporterScore" class="external-link"><img src="http://cranlogs.r-pkg.org/badges/grand-total/ReporterScore"></a> <a href="https://cran.r-project.org/package=ReporterScore" class="external-link"><img src="http://cranlogs.r-pkg.org/badges/last-month/ReporterScore"></a> <a href="https://cran.r-project.org/package=ReporterScore" class="external-link"><img src="https://www.r-pkg.org/badges/version/ReporterScore?color=green"></a> <a href="https://github.com/Asa12138/ReporterScore" class="external-link"><img src="https://img.shields.io/badge/devel%20version-0.1.6-green.svg"></a> <!-- badges: end --></p>
<p>受经典RSA启发，我们开发了改进的广义报告评分分析（GRSA）方法，并在R包ReporterScore中实现了该方法，还集成了全面的可视化方法和通路数据库。</p>
<p>GRSA是一种无阈值方法，适用于所有类型的生物医学特征，如基因、化合物和微生物物种。GRSA可在mixed模式（经典RSA）和directed模式（增强RSA）下工作。directed模式使用报告评分的符号区分上调或下调的通路，所以更直观。</p>
<p>重要的是，GRSA支持多组和纵向实验设计，因为它包含了与多组和纵向实验设计兼容的统计方法。 ReporterScore包还支持自定义的分层和关系数据库，为高级用户提供额外的灵活性。</p>
<p><img src="reference/figures/1-workflow.png" width="100%"></p>
<p>最新版本的HTML文档可在<a href="https://asa12138.github.io/ReporterScore/" class="external-link">Github 主页</a>找到.</p>
<div class="section level2">
<h2 id="引用">引用<a class="anchor" aria-label="anchor" href="#%E5%BC%95%E7%94%A8"></a></h2>
<p>在出版物中引用ReporterScore：</p>
<p>C. Peng, Q. Chen, S. Tan, X. Shen, C. Jiang, Generalized Reporter Score-based Enrichment Analysis for Omics Data. <em>Briefings in Bioinformatics</em> (2024). <a href="https://doi.org/10.1093/bib/bbae116" class="external-link uri">https://doi.org/10.1093/bib/bbae116</a>.</p>
</div>
<div class="section level2">
<h2 id="安装">安装<a class="anchor" aria-label="anchor" href="#%E5%AE%89%E8%A3%85"></a></h2>
<p>您可以从<a href="https://CRAN.R-project.org" class="external-link">CRAN</a>安装已发布版本的<code>ReporterScore</code>：</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"ReporterScore"</span><span class="op">)</span></span></code></pre></div>
<p>也可以从<a href="https://github.com/" class="external-link">GitHub</a>安装开发版本 <code>ReporterScore</code>：</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Asa12138/pcutils"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"Asa12138/ReporterScore"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="使用">使用<a class="anchor" aria-label="anchor" href="#%E4%BD%BF%E7%94%A8"></a></h2>
<div class="section level3">
<h3 id="id_1-输入数据特征丰度表和元数据">1. 输入数据（特征丰度表和元数据）<a class="anchor" aria-label="anchor" href="#id_1-%E8%BE%93%E5%85%A5%E6%95%B0%E6%8D%AE%E7%89%B9%E5%BE%81%E4%B8%B0%E5%BA%A6%E8%A1%A8%E5%92%8C%E5%85%83%E6%95%B0%E6%8D%AE"></a></h3>
<ul><li>对于特定物种的转录组、单细胞RNA-seq和相关基因组学数据，可以使用完整的基因丰度表。</li>
<li>对于涉及多种不同物种的宏基因组和宏转录组数据，可以使用KO丰度表，使用Blast、Diamond或KEGG官方映射软件来将reads或contigs对齐到KEGG或EggNOG数据库。</li>
<li>对于代谢组数据，可以使用已注释的化合物丰度表，但需要进行化合物ID的标准化（例如，将化合物ID转换为KEGG数据库中的C编号）。</li>
</ul><div class="section level4">
<h4 id="丰度表的格式">丰度表的格式：<a class="anchor" aria-label="anchor" href="#%E4%B8%B0%E5%BA%A6%E8%A1%A8%E7%9A%84%E6%A0%BC%E5%BC%8F"></a></h4>
<p>⚠️ <strong>重要提示：输入的丰度表不应进行预过滤（保留背景信息），因为‘GRSA’是一种无阈值的方法。</strong></p>
<ul><li>行名为特征ID（例如，“K00001”（KEGG K号）如果特征为“ko”；“PEX11A”（基因符号）如果特征为“gene”；“C00024”（KEGG C号）如果特征为“compound”）。</li>
<li>列名为样本。</li>
<li>丰度值可以是读数count或标准化值（例如TPM、FPKM、RPKM或相对丰度），对应适当的统计检验方法。</li>
</ul><p>适用于KO丰度表的示例代码如下：</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"KO_abundance_test"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">KO_abundance</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;                WT1         WT2         WT3         WT4         WT5         WT6</span></span>
<span><span class="co">#&gt; K03169 0.002653545 0.005096380 0.002033923 0.000722349 0.003468322 0.001483028</span></span>
<span><span class="co">#&gt; K07133 0.000308237 0.000280458 0.000596527 0.000859854 0.000308719 0.000878098</span></span>
<span><span class="co">#&gt; K03088 0.002147068 0.002030742 0.003797459 0.004161979 0.002076596 0.003091182</span></span>
<span><span class="co">#&gt; K03530 0.003788366 0.000239298 0.000445817 0.000557271 0.000222969 0.000529624</span></span>
<span><span class="co">#&gt; K06147 0.000785654 0.001213630 0.001312569 0.001662740 0.002387006 0.001725797</span></span>
<span><span class="co">#&gt; K05349 0.001816325 0.002813642 0.003274701 0.001089906 0.002371921 0.001795214</span></span></code></pre></div>
<p>你还应该提供一个实验设计元数据：</p>
</div>
<div class="section level4">
<h4 id="元数据表的格式">元数据表的格式：<a class="anchor" aria-label="anchor" href="#%E5%85%83%E6%95%B0%E6%8D%AE%E8%A1%A8%E7%9A%84%E6%A0%BC%E5%BC%8F"></a></h4>
<ul><li>行名为样本，列名为分组。</li>
<li>分组变量可以是类别（至少包含两个类别，用于差异丰度分析）。</li>
<li>分组变量也可以是多个时间点（用于纵向分析）。</li>
<li>分组变量也可以是连续变量（用于相关性分析）。</li>
</ul><div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="co">#&gt;     Group Group2</span></span>
<span><span class="co">#&gt; WT1    WT     G3</span></span>
<span><span class="co">#&gt; WT2    WT     G3</span></span>
<span><span class="co">#&gt; WT3    WT     G3</span></span>
<span><span class="co">#&gt; WT4    WT     G3</span></span>
<span><span class="co">#&gt; WT5    WT     G3</span></span>
<span><span class="co">#&gt; WT6    WT     G1</span></span></code></pre></div>
<p>⚠️ <strong>重要提示：元数据的行名和特征丰度表的列名应该匹配或部分匹配！</strong></p>
<p><code>ReporterScore</code> 将根据元数据的行名和特征丰度表的列名自动进行样本匹配。</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">KO_abundance</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">KO_abundance</span><span class="op">)</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">0</span></span>
<span><span class="co">## TRUE</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="id_2-通路数据库">2. 通路数据库<a class="anchor" aria-label="anchor" href="#id_2-%E9%80%9A%E8%B7%AF%E6%95%B0%E6%8D%AE%E5%BA%93"></a></h3>
<p><code>ReporterScore</code> 包含内置的KEGG通路、模块、基因、化合物和GO数据库，并支持自定义数据库，使其能够兼容来自多种组学数据的特征丰度表。</p>
<p><strong>你可以根据自己的分析需求选择以下任一方法来加载数据库：</strong></p>
<ol style="list-style-type: decimal"><li><p>对于KO丰度表，<code>ReporterScore</code> 包含内置的KEGG通路-KO和模块-KO数据库（2023-08版本）。您可以使用 <code><a href="reference/load_htable.html">load_KOlist()</a></code> 查看这些数据库，并使用 <code><a href="reference/update_KEGG.html">update_KO_file()</a></code> 更新这些数据库（通过KEGG API），保持数据库的最新性非常重要。</p></li>
<li><p>对于化合物丰度表，<code>ReporterScore</code> 包含内置的KEGG通路-化合物和模块-化合物数据库（2023-08版本）。您可以使用 <code><a href="reference/load_htable.html">load_CPDlist()</a></code> 查看这些数据库，并使用 <code><a href="reference/update_KEGG.html">update_KO_file()</a></code> 更新这些数据库（通过KEGG API）。</p></li>
<li><p>对于KO、基因和化合物丰度表，<code>ReporterScore</code> 包含人类（hsa）和小鼠（mmu）的内置通路-ko、通路-基因和通路-化合物数据库。您可以使用 <code><a href="reference/custom_modulelist_from_org.html">custom_modulelist_from_org()</a></code> 查看这些数据库。使用 <code><a href="reference/update_KEGG.html">update_org_pathway()</a></code> 更新这些数据库，并下载其他生物数据库（通过KEGG API）。</p></li>
<li><p><code>ReporterScore</code> 包含内置的GO-基因数据库。您可以使用 <code><a href="reference/load_GOlist.html">load_GOlist()</a></code> 查看这些数据库，并使用 <code><a href="reference/update_GOlist.html">update_GOlist()</a></code> 更新这些数据库（通过KEGG API）。</p></li>
<li><p>您可以通过使用 <code><a href="reference/custom_modulelist.html">custom_modulelist()</a></code> 自定义您自己的通路数据库（感兴趣的基因集）。</p></li>
</ol><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1. KEGG pathway-KO and module-KO databases</span></span>
<span><span class="va">KOlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_htable.html">load_KOlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">KOlist</span><span class="op">$</span><span class="va">pathway</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 2. KEGG pathway-compound and module-compound databases</span></span>
<span><span class="va">CPDlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_htable.html">load_CPDlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CPDlist</span><span class="op">$</span><span class="va">pathway</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. human (hsa) pathway-ko/gene/compound databases</span></span>
<span><span class="va">hsa_pathway_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/custom_modulelist_from_org.html">custom_modulelist_from_org</a></span><span class="op">(</span></span>
<span>  org <span class="op">=</span> <span class="st">"hsa"</span>,</span>
<span>  feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ko"</span>, <span class="st">"gene"</span>, <span class="st">"compound"</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">hsa_pathway_gene</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. GO-gene database</span></span>
<span><span class="va">GOlist</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_GOlist.html">load_GOlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">GOlist</span><span class="op">$</span><span class="va">BP</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 5. 自定义自己的通路数据库</span></span>
<span><span class="op">?</span><span class="fu"><a href="reference/custom_modulelist.html">custom_modulelist</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="id_3-一步grsa">3. 一步GRSA<a class="anchor" aria-label="anchor" href="#id_3-%E4%B8%80%E6%AD%A5grsa"></a></h3>
<p>使用函数 <code>GRSA</code> 或 <code>reporter_score</code> 可以一步得到Reporter Score结果。</p>
<p>⚠️分析中有一些重要的参数：</p>
<ul><li>
<strong>mode</strong>: “mixed” 或 “directed”（仅用于两组差异分析或多组相关分析。详情见 <code>pvalue2zs</code>）。</li>
<li>
<strong>method</strong>: 计算p值的统计测试方法。默认为 <code>wilcox.test</code>：
<ul><li>
<code>t.test</code>（参数检验）和 <code>wilcox.test</code>（非参数检验）。用于比较两组样本。</li>
<li>
<code>anova</code>（参数检验）和 <code>kruskal.test</code>（非参数检验）。执行单因素方差分析或Kruskal-Wallis秩和检验，比较多个组。</li>
<li>“pearson”、“kendall” 或 “spearman”（相关性检验），参见 <code>cor</code>。</li>
<li>“none”: 使用“none”进行“逐步富集”计算p值。您可以使用“DESeq2”、“Edger”、“Limma”、“ALDEX”、“ANCOM”等其他方法自行计算p值。</li>
</ul></li>
<li>
<strong>type</strong>: 选择内置的通路数据库：
<ul><li>‘pathway’ 或 ‘module’ 默认KEGG数据库适用于<strong>微生物组</strong>。</li>
<li>‘CC’, ‘MF’, ‘BP’, ‘ALL’ 默认GO数据库适用于<strong>人类</strong>。</li>
<li>org 列在 <a href="https://www.genome.jp/kegg/catalog/org_list.html" class="external-link uri">https://www.genome.jp/kegg/catalog/org_list.html</a> 中，例如 ‘hsa’（如果您的kodf来自特定的生物体，您应在此处指定type）。</li>
</ul></li>
<li>
<strong>modulelist</strong>: 自定义数据库。包含 ‘id’,‘K_num’,‘KOs’,‘Description’ 列的数据框。以 <code>KOlist</code> 为例，使用 <code>custom_modulelist</code> 构建自定义数据库。</li>
<li>
<strong>feature</strong>: “ko”, “gene”, “compound” 中的一个。</li>
</ul><p><strong>分组中第一个将被设置为对照组，您可以更改因子水平以改变比较。</strong></p>
<p>例如，我们想比较两组 ‘WT-OE’，并使用“directed”模式，因为我们只需要知道<strong>OE组</strong>中哪些通路上调或下调（反过来就是WT组通路下调或上调）：</p>
<div class="section level4">
<h4 id="ko-pathway">KO-pathway<a class="anchor" aria-label="anchor" href="#ko-pathway"></a></h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Comparison: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Comparison:  WT OE</span></span>
<span></span>
<span><span class="co"># for microbiome!!!</span></span>
<span><span class="va">reporter_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"pathway"</span>, feature <span class="op">=</span> <span class="st">"ko"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ================================Use feature: ko=================================</span></span>
<span><span class="co">#&gt; ===============================Checking rownames================================</span></span>
<span><span class="co">#&gt; Some of your ko_stat are not KO id, check the format! (e.g. K00001)</span></span>
<span><span class="co">#&gt; 52.7% of your kos in the modulelist!</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ===================================1.KO test====================================</span></span>
<span><span class="co">#&gt; =================================Checking group=================================</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ==============================Calculating each KO===============================</span></span>
<span><span class="co">#&gt; ===========================Using method: wilcox.test============================</span></span>
<span><span class="co">#&gt; 1000 features done.</span></span>
<span><span class="co">#&gt; 2000 features done.</span></span>
<span><span class="co">#&gt; 3000 features done.</span></span>
<span><span class="co">#&gt; 4000 features done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Compared groups: WT, OE</span></span>
<span><span class="co">#&gt; Total KO number: 4535</span></span>
<span><span class="co">#&gt; Compare method: wilcox.test</span></span>
<span><span class="co">#&gt; Time use: 1.179</span></span>
<span><span class="co">#&gt; =========================2.Transfer p.value to Z-score==========================</span></span>
<span><span class="co">#&gt; ==========================3.Calculating reporter score==========================</span></span>
<span><span class="co">#&gt; ==================================load KOlist===================================</span></span>
<span><span class="co">#&gt; ===================KOlist download time: 2023-08-14 16:00:52====================</span></span>
<span><span class="co">#&gt; If you want to update KOlist, use `update_KO_file()`</span></span>
<span><span class="co">#&gt; ============================Calculating each pathway============================</span></span>
<span><span class="co">#&gt; 100 pathways done.</span></span>
<span><span class="co">#&gt; 400 pathways done.</span></span>
<span><span class="co">#&gt; ID number: 481</span></span>
<span><span class="co">#&gt; Time use: 1.689</span></span>
<span><span class="co">#&gt; ====================================All done====================================</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="gene-pathway">Gene-pathway<a class="anchor" aria-label="anchor" href="#gene-pathway"></a></h4>
<p>当您使用特定物种（例如人类）的基因丰度表时，请记得设置 <code>feature</code> 和 <code>type</code> 参数！或者通过 <code>modulelist</code> 提供数据库：</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"genedf"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 方法 1: 设置 `feature` 和 `type` 参数</span></span>
<span><span class="va">reporter_res_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">genedf</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"hsa"</span>, feature <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ===============================Use feature: gene================================</span></span>
<span><span class="co">#&gt; ===============================Checking rownames================================</span></span>
<span><span class="co">#&gt; please make sure your input table rows are gene symbol!</span></span>
<span><span class="co">#&gt; 100% of your genes in the modulelist!</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ===================================1.KO test====================================</span></span>
<span><span class="co">#&gt; =================================Checking group=================================</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ==============================Calculating each KO===============================</span></span>
<span><span class="co">#&gt; ===========================Using method: wilcox.test============================</span></span>
<span><span class="co">#&gt; 1000 features done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Compared groups: WT, OE</span></span>
<span><span class="co">#&gt; Total KO number: 1000</span></span>
<span><span class="co">#&gt; Compare method: wilcox.test</span></span>
<span><span class="co">#&gt; Time use: 0.224</span></span>
<span><span class="co">#&gt; =========================2.Transfer p.value to Z-score==========================</span></span>
<span><span class="co">#&gt; ==========================3.Calculating reporter score==========================</span></span>
<span><span class="co">#&gt; ================================load hsa pathway================================</span></span>
<span><span class="co">#&gt; =================hsa pathway download time: 2023-08-14 23:28:13=================</span></span>
<span><span class="co">#&gt; If you want to update hsa pathway, use `update_org_pathway('hsa')`</span></span>
<span><span class="co">#&gt; please assgin this custom modulelist to `reporter_score(modulelist=your_modulelist)` to do a custom enrichment.</span></span>
<span><span class="co">#&gt; You choose the feature: 'gene', make sure the rownames of your input table are right.</span></span>
<span><span class="co">#&gt; ============================Calculating each pathway============================</span></span>
<span><span class="co">#&gt; 100 pathways done.</span></span>
<span><span class="co">#&gt; 150 pathways done.</span></span>
<span><span class="co">#&gt; 200 pathways done.</span></span>
<span><span class="co">#&gt; 250 pathways done.</span></span>
<span><span class="co">#&gt; 300 pathways done.</span></span>
<span><span class="co">#&gt; ID number: 343</span></span>
<span><span class="co">#&gt; Time use: 1.558</span></span>
<span><span class="co">#&gt; ====================================All done====================================</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 方法 2: 通过`modulelist`给出数据库，结果与方法1相同.</span></span>
<span><span class="va">hsa_pathway_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/custom_modulelist_from_org.html">custom_modulelist_from_org</a></span><span class="op">(</span></span>
<span>  org <span class="op">=</span> <span class="st">"hsa"</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"gene"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">reporter_res_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">genedf</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  modulelist <span class="op">=</span> <span class="va">hsa_pathway_gene</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res_gene</span>, rs_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 使用`modify_description`删除通路描述的后缀</span></span>
<span><span class="va">reporter_res_gene2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/modify_description.html">modify_description</a></span><span class="op">(</span><span class="va">reporter_res_gene</span>, pattern <span class="op">=</span> <span class="st">" - Homo sapiens (human)"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res_gene2</span>, rs_threshold <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 使用“ggplot_translator”翻译通路描述</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">pcutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/ggplot_translator.html" class="external-link">ggplot_translator</a></span><span class="op">(</span><span class="va">p2</span><span class="op">)</span></span>
<span><span class="co">#&gt; Please set the font family to make the labels display well.</span></span>
<span><span class="co">#&gt;  see `how_to_set_font_for_plot()`.</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">/</span> <span class="va">p2</span> <span class="op">/</span> <span class="va">p3</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-10-1.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="compound-pathway">Compound-pathway<a class="anchor" aria-label="anchor" href="#compound-pathway"></a></h4>
<p>当您使用化合物丰度表时，请记得设置 <code>feature</code> 和 <code>type</code> 参数！或者通过 <code>modulelist</code> 提供数据库。</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reporter_res_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">chem_df</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"wilcox.test"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"hsa"</span>, feature <span class="op">=</span> <span class="st">"compound"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="输出结果">输出结果<a class="anchor" aria-label="anchor" href="#%E8%BE%93%E5%87%BA%E7%BB%93%E6%9E%9C"></a></h4>
<p>结果是一个“reporter_score”对象：</p>
<table class="table"><thead><tr class="header"><th>元素</th>
<th>描述</th>
</tr></thead><tbody><tr class="odd"><td><code>kodf</code></td>
<td>您的输入KO丰度表</td>
</tr><tr class="even"><td><code>ko_stat</code></td>
<td>KO统计结果，包含p值和z分数</td>
</tr><tr class="odd"><td><code>reporter_s</code></td>
<td>每个通路中的报告者分数</td>
</tr><tr class="even"><td><code>modulelist</code></td>
<td>默认的KOlist或自定义的模块列表数据框</td>
</tr><tr class="odd"><td><code>group</code></td>
<td>您数据中的比较组</td>
</tr><tr class="even"><td><code>metadata</code></td>
<td>包含样本信息的数据框，包括组信息</td>
</tr></tbody></table><p>重要的结果是 <code>reporter_res$reporter_s</code>，它是一个包含每个通路中报告者分数的数据框：</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 在Rstudio中查看data.frame</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/View.html" class="external-link">View</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">$</span><span class="va">reporter_s</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 将结果导出为.csv格式，并使用Excel进行查看：</span></span>
<span><span class="fu"><a href="reference/export_report_table.html">export_report_table</a></span><span class="op">(</span><span class="va">reporter_res</span>, dir_name <span class="op">=</span> <span class="st">"~/Downloads/"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="id_4-可视化">4. 可视化<a class="anchor" aria-label="anchor" href="#id_4-%E5%8F%AF%E8%A7%86%E5%8C%96"></a></h3>
<p>在获得Reporter Score结果后，我们可以通过多种方式进行可视化。</p>
<p>当我们关注整体结果时：</p>
<ul><li>绘制显著富集的通路：</li>
</ul><p>您可以设置 <code>rs_threshold</code> 来过滤部分通路， <strong>默认的 <code>rs_threshold</code> 是 1.64， 对应的显著性水平为 0.05</strong>。</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(reporter_res$reporter_s)</span></span>
<span><span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res</span>, rs_threshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span><span class="op">)</span>, facet_level <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; ==============================load Pathway_htable===============================</span></span>
<span><span class="co">#&gt; ===============Pathway_htable download time: 2024-01-12 00:52:39================</span></span>
<span><span class="co">#&gt; If you want to update Pathway_htable, use `update_htable(type='pathway')`</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-12-1.png" width="100%"></p>
<p>⚠️<strong>在directed模式下，富集在一个组中意味着在另一个组中下调。</strong></p>
<ul><li>绘制最显著富集的通路（珠包图）：</li>
</ul><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_report_circle_packing.html">plot_report_circle_packing</a></span><span class="op">(</span><span class="va">reporter_res</span>, rs_threshold <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2.5</span>, <span class="fl">2.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; ==============================load Pathway_htable===============================</span></span>
<span><span class="co">#&gt; ===============Pathway_htable download time: 2024-01-12 00:52:39================</span></span>
<span><span class="co">#&gt; If you want to update Pathway_htable, use `update_htable(type='pathway')`</span></span>
<span><span class="co">#&gt; Non-leaf weights ignored</span></span>
<span><span class="co">#&gt; Scale for fill is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for fill, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-13-1.png" width="100%"></p>
<p>当我们关注一条通路时，例如“map00780”：</p>
<ul><li>绘制箱线图：</li>
</ul><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_in_pathway.html">plot_KOs_in_pathway</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-14-1.png" width="100%"></p>
<ul><li>绘制Z分数的分布图：</li>
</ul><div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_distribution.html">plot_KOs_distribution</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-15-1.png" width="100%"></p>
<ul><li>绘制网络图：</li>
</ul><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_network.html">plot_KOs_network</a></span><span class="op">(</span><span class="va">reporter_res</span>,</span>
<span>  map_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"map00780"</span>, <span class="st">"map00785"</span>, <span class="st">"map00900"</span><span class="op">)</span>,</span>
<span>  main <span class="op">=</span> <span class="st">""</span>, mark_module <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-1.png" width="100%"></p>
<ul><li>在通路中绘制KOs丰度：</li>
</ul><div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_box.html">plot_KOs_box</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span>, only_sig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; `geom_smooth()` using formula = 'y ~ x'</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-17-1.png" width="100%"></p>
<ul><li>在通路中绘制KOs丰度（热图）：</li>
</ul><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_heatmap.html">plot_KOs_heatmap</a></span><span class="op">(</span><span class="va">reporter_res</span>,</span>
<span>  map_id <span class="op">=</span> <span class="st">"map00780"</span>, only_sig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  heatmap_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>cutree_rows <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-18-1.png" width="100%"></p>
<ul><li>绘制KEGG通路图：</li>
</ul><div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_KEGG_map.html">plot_KEGG_map</a></span><span class="op">(</span><span class="va">reporter_res</span>, map_id <span class="op">=</span> <span class="st">"map00780"</span>, color_var <span class="op">=</span> <span class="st">"Z_score"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/ko00780.Z_score.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="多组或纵向研究">多组或纵向研究<a class="anchor" aria-label="anchor" href="#%E5%A4%9A%E7%BB%84%E6%88%96%E7%BA%B5%E5%90%91%E7%A0%94%E7%A9%B6"></a></h3>
<p>如果我们的实验设计涉及多于两组或纵向研究，我们可以选择多组比较（或相关分析）：</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Comparison: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">$</span><span class="va">Group2</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Comparison:  G1 G2 G3</span></span>
<span></span>
<span><span class="va">reporter_res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group2"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"spearman"</span>, perm <span class="op">=</span> <span class="fl">999</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; ================================Use feature: ko=================================</span></span>
<span><span class="co">#&gt; ===============================Checking rownames================================</span></span>
<span><span class="co">#&gt; Some of your ko_stat are not KO id, check the format! (e.g. K00001)</span></span>
<span><span class="co">#&gt; 52.7% of your kos in the modulelist!</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ===================================1.KO test====================================</span></span>
<span><span class="co">#&gt; =================================Checking group=================================</span></span>
<span><span class="co">#&gt; 30 samples are matched for next step.</span></span>
<span><span class="co">#&gt; ===========================Removing all-zero rows: 0============================</span></span>
<span><span class="co">#&gt; ==============================Calculating each KO===============================</span></span>
<span><span class="co">#&gt; =============================Using method: spearman=============================</span></span>
<span><span class="co">#&gt; Using correlation analysis: spearman, the groups will be transform to numeric, note the factor feature of group.</span></span>
<span><span class="co">#&gt; 1000 features done.</span></span>
<span><span class="co">#&gt; 2000 features done.</span></span>
<span><span class="co">#&gt; 3000 features done.</span></span>
<span><span class="co">#&gt; 4000 features done.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Compared groups: G1, G2, G3</span></span>
<span><span class="co">#&gt; Total KO number: 4535</span></span>
<span><span class="co">#&gt; Compare method: spearman</span></span>
<span><span class="co">#&gt; Time use: 0.549</span></span>
<span><span class="co">#&gt; =========================2.Transfer p.value to Z-score==========================</span></span>
<span><span class="co">#&gt; ==========================3.Calculating reporter score==========================</span></span>
<span><span class="co">#&gt; ==================================load KOlist===================================</span></span>
<span><span class="co">#&gt; ===================KOlist download time: 2023-08-14 16:00:52====================</span></span>
<span><span class="co">#&gt; If you want to update KOlist, use `update_KO_file()`</span></span>
<span><span class="co">#&gt; ============================Calculating each pathway============================</span></span>
<span><span class="co">#&gt; 100 pathways done.</span></span>
<span><span class="co">#&gt; 400 pathways done.</span></span>
<span><span class="co">#&gt; ID number: 481</span></span>
<span><span class="co">#&gt; Time use: 1.643</span></span>
<span><span class="co">#&gt; ====================================All done====================================</span></span>
<span></span>
<span><span class="fu"><a href="reference/plot_features_in_pathway.html">plot_KOs_in_pathway</a></span><span class="op">(</span><span class="va">reporter_res2</span>, map_id <span class="op">=</span> <span class="st">"map02060"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-21-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="指定pattern">指定pattern<a class="anchor" aria-label="anchor" href="#%E6%8C%87%E5%AE%9Apattern"></a></h3>
<p>例如，如果我们期望出现指数增长趋势，可以将组别 “G1”，“G2” 和 “G3” 设置为 1、10 和 100。</p>
<p>我们使用 1、5、1 来寻找表现为下降-上升-下降模式的通路。</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reporter_res3</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/reporter_score.html">GRSA</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group2"</span>, <span class="va">metadata</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"directed"</span>, perm <span class="op">=</span> <span class="fl">999</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pearson"</span>, pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"G1"</span> <span class="op">=</span> <span class="fl">1</span>, <span class="st">"G2"</span> <span class="op">=</span> <span class="fl">5</span>, <span class="st">"G3"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">reporter_res3</span>, rs_threshold <span class="op">=</span> <span class="fl">3</span>, show_ID <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-1.png" width="100%"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="reference/plot_features_in_pathway.html">plot_KOs_in_pathway</a></span><span class="op">(</span><span class="va">reporter_res3</span>, map_id <span class="op">=</span> <span class="st">"map00860"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-2.png" width="100%"></p>
<p>为了探索数据中的潜在patterns，使用聚类方法，如C均值聚类。</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rsa_cm_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/RSA_by_cm.html">RSA_by_cm</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group2"</span>, <span class="va">metadata</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  k_num <span class="op">=</span> <span class="fl">3</span>, perm <span class="op">=</span> <span class="fl">999</span></span>
<span><span class="op">)</span></span>
<span><span class="co"># show the patterns</span></span>
<span><span class="fu"><a href="reference/RSA_by_cm.html">plot_c_means</a></span><span class="op">(</span><span class="va">rsa_cm_res</span>, filter_membership <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-23-1.png" width="100%"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="reference/plot_report.html">plot_report_bar</a></span><span class="op">(</span><span class="va">rsa_cm_res</span>, rs_threshold <span class="op">=</span> <span class="fl">2.5</span>, y_text_size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-23-2.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="详细信息">详细信息<a class="anchor" aria-label="anchor" href="#%E8%AF%A6%E7%BB%86%E4%BF%A1%E6%81%AF"></a></h2>
<div class="section level3">
<h3 id="逐步说明">逐步说明<a class="anchor" aria-label="anchor" href="#%E9%80%90%E6%AD%A5%E8%AF%B4%E6%98%8E"></a></h3>
<p>单步函数 <code>reporter_score</code>/<code>GRSA</code> 包含三个部分：</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"KO_abundance_test"</span><span class="op">)</span></span>
<span><span class="va">ko_pvalue</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ko.test.html">ko.test</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>, method <span class="op">=</span> <span class="st">"wilcox.test"</span><span class="op">)</span></span>
<span><span class="va">ko_stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/pvalue2zs.html">pvalue2zs</a></span><span class="op">(</span><span class="va">ko_pvalue</span>, mode <span class="op">=</span> <span class="st">"directed"</span><span class="op">)</span></span>
<span><span class="va">reporter_s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_reporter_score.html">get_reporter_score</a></span><span class="op">(</span><span class="va">ko_stat</span>, perm <span class="op">=</span> <span class="fl">499</span><span class="op">)</span></span></code></pre></div>
<ol style="list-style-type: decimal"><li><p><code>ko.test</code>: 此函数用于计算 KO_abundance 的 <em>p-value</em>，可使用各种内置方法，如差异分析 (<code>t.test</code>, <code>wilcox.test</code>, <code>kruskal.test</code>, <code>anova</code>) 或相关分析 (<code>pearson</code>, <code>spearman</code>, <code>kendall</code>)。<strong>您也可以使用其他方法计算 KO_abundance 的 <em>p-value</em>，比如 “DESeq2”, “Edger”, “Limma”, “ALDEX”, “ANCOM”，并自行进行 p.adjust，然后跳过 <code>ko.test</code> 步骤，继续下一步……</strong></p></li>
<li><p><code>pvalue2zs</code>: 此函数将 KOs 的 <em>p-value</em> 转换为 Z 分数（选择模式: “mixed” 或 “directed”）。</p></li>
<li><p><code>get_reporter_score</code>: 此函数在特定数据库中计算每个通路的Reporter Score。您可以在此处使用自定义数据库。</p></li>
</ol><p>以 “Limma” 为例：</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 1-1. 使用 Limma 计算 p-value</span></span>
<span><span class="va">ko_pvalue</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/ko.test.html">ko.test</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>, method <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ko_Limma_p</span> <span class="op">&lt;-</span> <span class="fu">pctax</span><span class="fu">::</span><span class="fu">diff_da</span><span class="op">(</span><span class="va">KO_abundance</span>, group_df <span class="op">=</span> <span class="va">metadata</span><span class="op">[</span><span class="st">"Group"</span><span class="op">]</span>, method <span class="op">=</span> <span class="st">"limma"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1-2. 替换 ko_pvalue 中的 p-value，记得匹配 KO_ids</span></span>
<span><span class="va">ko_pvalue</span><span class="op">$</span><span class="va">`p.value`</span> <span class="op">&lt;-</span> <span class="va">ko_Limma_p</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ko_pvalue</span><span class="op">$</span><span class="va">KO_id</span>, <span class="va">ko_Limma_p</span><span class="op">$</span><span class="va">tax</span><span class="op">)</span>, <span class="st">"pvalue"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># 2. 使用 `pvalue2zs` 获取 Z 分数</span></span>
<span><span class="va">ko_stat</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/pvalue2zs.html">pvalue2zs</a></span><span class="op">(</span><span class="va">ko_pvalue</span>, mode <span class="op">=</span> <span class="st">"directed"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. 使用 `get_reporter_score` 获取Reporter Score</span></span>
<span><span class="va">reporter_s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_reporter_score.html">get_reporter_score</a></span><span class="op">(</span><span class="va">ko_stat</span>, perm <span class="op">=</span> <span class="fl">499</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 4. 组合结果</span></span>
<span><span class="va">reporter_res1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/combine_rs_res.html">combine_rs_res</a></span><span class="op">(</span><span class="va">KO_abundance</span>, <span class="st">"Group"</span>, <span class="va">metadata</span>, <span class="va">ko_stat</span>, <span class="va">reporter_s1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 然后可以使用 reporter_res1 进行可视化</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="其他常用的富集方法">其他常用的富集方法<a class="anchor" aria-label="anchor" href="#%E5%85%B6%E4%BB%96%E5%B8%B8%E7%94%A8%E7%9A%84%E5%AF%8C%E9%9B%86%E6%96%B9%E6%B3%95"></a></h3>
<table class="table"><colgroup><col width="3%"><col width="23%"><col width="17%"><col width="55%"></colgroup><thead><tr class="header"><th align="left">Category</th>
<th align="left">Method</th>
<th align="left">Tools</th>
<th align="left">Notes</th>
</tr></thead><tbody><tr class="odd"><td align="left">ORA</td>
<td align="left">Hypergeometric test / Fisher’s exact test</td>
<td align="left">DAVID (website) , clusterProfiler (R package)</td>
<td align="left">The most common methods used in enrichment analysis. Selecting a list of genes is required.</td>
</tr><tr class="even"><td align="left">FCS</td>
<td align="left">Gene set enrichment analysis (GSEA)</td>
<td align="left">GSEA (website)</td>
<td align="left">GSEA creatively uses gene ranking, rather than selecting a list of genes, to identify statistically significant and concordant differences across gene sets.</td>
</tr><tr class="odd"><td align="left">FCS</td>
<td align="left">Generalized reporter score-based analysis (GRSA/RSA)</td>
<td align="left">ReporterScore (R package developed in this study)</td>
<td align="left">Find significant metabolites (first report), pathways, and taxonomy based on the p-values for multi-omics data.</td>
</tr><tr class="even"><td align="left">FCS</td>
<td align="left">Significance Analysis of Function and Expression (SAFE)</td>
<td align="left">safe (R package)</td>
<td align="left">SAFE assesses the significance of gene categories by calculating both local and global statistics from gene expression data.</td>
</tr><tr class="odd"><td align="left">FCS</td>
<td align="left">Gene Set Analysis (GSA)</td>
<td align="left">GSA (R Package)</td>
<td align="left">GSA was proposed as an improvement of GSEA, using the “maxmean” statistic instead of the weighted sign KS statistic.</td>
</tr><tr class="even"><td align="left">FCS</td>
<td align="left">Pathway Analysis with Down-weighting of Overlapping Genes (PADOG)</td>
<td align="left">PADOG (R package)</td>
<td align="left">PADOGA assumes that genes associated with fewer pathways have more significant effects than genes associated with more pathways.</td>
</tr><tr class="odd"><td align="left">FCS</td>
<td align="left">Gene Set Variation Analysis (GSVA)</td>
<td align="left">GSVA (R package)</td>
<td align="left">A nonparametric, unsupervised method that transforms gene expression data into gene set scores for downstream differential pathway activity analysis.</td>
</tr><tr class="even"><td align="left">PT</td>
<td align="left">Topology-based pathway enrichment analysis (TPEA)</td>
<td align="left">TPEA (R package)</td>
<td align="left">Integrate topological properties and global upstream/downstream positions of genes in pathways.</td>
</tr></tbody></table><p>组学数据常用的富集方法.</p>
<p><code>ReporterScore</code> 还提供其他的富集方法，如 <code>KO_fisher</code>（fisher.test）、<code>KO_enrich</code>（fisher.test，来自 <code>clusterProfiler</code>）、<code>KO_gsea</code>（GSEA，来自 <code>clusterProfiler</code>）、<code>KO_gsa</code>（GSA，来自 <code>GSA</code>）、<code>KO_safe</code>（SAFE，来自 <code>safe</code>）、<code>KO_padog</code>（PADOG，来自 <code>PADOG</code>）、<code>KO_gsva</code>（GSVA，来自 <code>GSVA</code>）。</p>
<p>输入数据来自 <code>reporter_score</code>，同时支持自定义数据库，因此您可以轻松比较各种富集方法的结果，并进行全面的分析：</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View(reporter_res2$reporter_s)</span></span>
<span><span class="co"># reporter_score</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">$</span><span class="va">reporter_s</span>, <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">ReporterScore</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1.64</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">RS</span></span>
<span><span class="co"># fisher</span></span>
<span><span class="va">fisher_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/KO_fisher.html">KO_fisher</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">fisher_res</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">Fisher</span></span>
<span><span class="co"># enricher</span></span>
<span><span class="va">enrich_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/KO_enrich.html">KO_enrich</a></span><span class="op">(</span><span class="va">reporter_res</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">enrich_res</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">clusterProfiler</span></span>
<span><span class="co"># GESA</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">gsea_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/KO_gsea.html">KO_gsea</a></span><span class="op">(</span><span class="va">reporter_res</span>, weight <span class="op">=</span> <span class="st">"Z_score"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, : There are ties in the preranked stats (57.15% of the list).</span></span>
<span><span class="co">#&gt; The order of those tied genes will be arbitrary, which may produce unexpected results.</span></span>
<span><span class="co">#&gt; Warning in fgseaMultilevel(pathways = pathways, stats = stats, minSize =</span></span>
<span><span class="co">#&gt; minSize, : For some pathways, in reality P-values are less than 1e-10. You can</span></span>
<span><span class="co">#&gt; set the `eps` argument to zero for better estimation.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">gsea_res</span><span class="op">)</span>, <span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="reference/pipe.html">%&gt;%</a></span> <span class="fu">pull</span><span class="op">(</span><span class="va">ID</span><span class="op">)</span> <span class="op">-&gt;</span> <span class="va">GSEA</span></span>
<span></span>
<span><span class="va">venn_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>GRSA <span class="op">=</span> <span class="va">RS</span>, Fisher <span class="op">=</span> <span class="va">Fisher</span>, CP <span class="op">=</span> <span class="va">clusterProfiler</span>, GSEA <span class="op">=</span> <span class="va">GSEA</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Asa12138/pcutils" class="external-link">pcutils</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/venn.html" class="external-link">venn</a></span><span class="op">(</span><span class="va">venn_res</span>, <span class="st">"network"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-25-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="其他功能">其他功能<a class="anchor" aria-label="anchor" href="#%E5%85%B6%E4%BB%96%E5%8A%9F%E8%83%BD"></a></h2>
<div class="section level3">
<h3 id="kos高层级整合">KOs高层级整合<a class="anchor" aria-label="anchor" href="#kos%E9%AB%98%E5%B1%82%E7%BA%A7%E6%95%B4%E5%90%88"></a></h3>
<p><a href="https://www.genome.jp/kegg/brite.html" class="external-link">KEGG BRITE</a> 是一个层次分类系统的集合，捕捉了各种生物对象的功能层次结构，特别是那些作为 KEGG 对象表示的对象。</p>
<p>我们收集了 k00001 KEGG Orthology (KO) 表，因此您可以总结每个层级的丰度。使用 <code>load_KO_htable</code> 获取 KO_htable，并使用 <code>update_KO_htable</code> 进行更新。使用 <code>up_level_KO</code> 可以将 KOs 升级到特定的层级，例如 “pathway”、”module”、”level1”、”level2”、”level3”、“module1”、“module2”、“module3”。</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">KO_htable</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_htable.html">load_KO_htable</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load KO_htable=================================</span></span>
<span><span class="co">#&gt; ==================KO_htable download time: 2024-01-12 00:49:03==================</span></span>
<span><span class="co">#&gt; If you want to update KO_htable, use `update_htable(type='ko')`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">KO_htable</span><span class="op">)</span></span>
<span><span class="co">#&gt;   level1_name             level2_name level3_id                  level3_name</span></span>
<span><span class="co">#&gt; 1  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 2  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 3  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 4  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 5  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt; 6  Metabolism Carbohydrate metabolism  map00010 Glycolysis / Gluconeogenesis</span></span>
<span><span class="co">#&gt;    KO_id                                                    KO_name</span></span>
<span><span class="co">#&gt; 1 K00844                                HK; hexokinase [EC:2.7.1.1]</span></span>
<span><span class="co">#&gt; 2 K12407                              GCK; glucokinase [EC:2.7.1.2]</span></span>
<span><span class="co">#&gt; 3 K00845                              glk; glucokinase [EC:2.7.1.2]</span></span>
<span><span class="co">#&gt; 4 K25026                              glk; glucokinase [EC:2.7.1.2]</span></span>
<span><span class="co">#&gt; 5 K01810       GPI, pgi; glucose-6-phosphate isomerase [EC:5.3.1.9]</span></span>
<span><span class="co">#&gt; 6 K06859 pgi1; glucose-6-phosphate isomerase, archaeal [EC:5.3.1.9]</span></span>
<span><span class="fu"><a href="reference/plot_htable.html">plot_htable</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"ko"</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load KO_htable=================================</span></span>
<span><span class="co">#&gt; ==================KO_htable download time: 2024-01-12 00:49:03==================</span></span>
<span><span class="co">#&gt; If you want to update KO_htable, use `update_htable(type='ko')`</span></span>
<span><span class="co">#&gt; Warning: Vectorized input to `element_text()` is not officially supported.</span></span>
<span><span class="co">#&gt; ℹ Results may be unexpected or may change in future versions of ggplot2.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-26-1.png" width="100%"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">KO_level1</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/up_level_KO.html">up_level_KO</a></span><span class="op">(</span><span class="va">KO_abundance</span>, level <span class="op">=</span> <span class="st">"level1"</span>, show_name <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load KO_htable=================================</span></span>
<span><span class="co">#&gt; ==================KO_htable download time: 2024-01-12 00:49:03==================</span></span>
<span><span class="co">#&gt; If you want to update KO_htable, use `update_htable(type='ko')`</span></span>
<span><span class="fu">pcutils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pcutils/man/stackplot.html" class="external-link">stackplot</a></span><span class="op">(</span><span class="va">KO_level1</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">KO_level1</span><span class="op">)</span> <span class="op">==</span> <span class="st">"Unknown"</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggsci</span><span class="fu">::</span><span class="fu">scale_fill_d3</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-27-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="card-for-args">CARD for ARGs<a class="anchor" aria-label="anchor" href="#card-for-args"></a></h3>
<p>为了方便，我还整合了CARD数据库的描述信息 <a href="https://card.mcmaster.ca/download/0/broadstreet-v3.2.8.tar.bz2" class="external-link uri">https://card.mcmaster.ca/download/0/broadstreet-v3.2.8.tar.bz2</a>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">CARDinfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/load_CARDinfo.html">load_CARDinfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; =================================load CARDinfo==================================</span></span>
<span><span class="co">#&gt; ==================CARDinfo download time: 2024-01-12 01:12:11===================</span></span>
<span><span class="co">#&gt; If you want to update CARDinfo, use `update_GOlist()`</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">CARDinfo</span><span class="op">$</span><span class="va">ARO_index</span><span class="op">)</span></span>
<span><span class="co">#&gt;         ARO Accession CVTERM ID Model Sequence ID Model ID</span></span>
<span><span class="co">#&gt; 3005099   ARO:3005099     43314              6143     3831</span></span>
<span><span class="co">#&gt; 3002523   ARO:3002523     38923              8144     1781</span></span>
<span><span class="co">#&gt; 3002524   ARO:3002524     38924                85      746</span></span>
<span><span class="co">#&gt; 3002525   ARO:3002525     38925              4719     1246</span></span>
<span><span class="co">#&gt; 3002526   ARO:3002526     38926               228     1415</span></span>
<span><span class="co">#&gt; 3002527   ARO:3002527     38927              5510     2832</span></span>
<span><span class="co">#&gt;                                                     Model Name</span></span>
<span><span class="co">#&gt; 3005099 23S rRNA (adenine(2058)-N(6))-methyltransferase Erm(A)</span></span>
<span><span class="co">#&gt; 3002523                                             AAC(2')-Ia</span></span>
<span><span class="co">#&gt; 3002524                                             AAC(2')-Ib</span></span>
<span><span class="co">#&gt; 3002525                                             AAC(2')-Ic</span></span>
<span><span class="co">#&gt; 3002526                                             AAC(2')-Id</span></span>
<span><span class="co">#&gt; 3002527                                             AAC(2')-Ie</span></span>
<span><span class="co">#&gt;                                                       ARO Name</span></span>
<span><span class="co">#&gt; 3005099 23S rRNA (adenine(2058)-N(6))-methyltransferase Erm(A)</span></span>
<span><span class="co">#&gt; 3002523                                             AAC(2')-Ia</span></span>
<span><span class="co">#&gt; 3002524                                             AAC(2')-Ib</span></span>
<span><span class="co">#&gt; 3002525                                             AAC(2')-Ic</span></span>
<span><span class="co">#&gt; 3002526                                             AAC(2')-Id</span></span>
<span><span class="co">#&gt; 3002527                                             AAC(2')-Ie</span></span>
<span><span class="co">#&gt;         Protein Accession DNA Accession                         AMR Gene Family</span></span>
<span><span class="co">#&gt; 3005099        AAB60941.1    AF002716.1 Erm 23S ribosomal RNA methyltransferase</span></span>
<span><span class="co">#&gt; 3002523        AAA03550.1      L06156.2                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002524        AAC44793.1      U41471.1                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002525        CCP42991.1    AL123456.3                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002526        AAB41701.1      U72743.1                                 AAC(2')</span></span>
<span><span class="co">#&gt; 3002527        CAC32082.1    AL583926.1                                 AAC(2')</span></span>
<span><span class="co">#&gt;                                                                   Drug Class</span></span>
<span><span class="co">#&gt; 3005099 lincosamide antibiotic;macrolide antibiotic;streptogramin antibiotic</span></span>
<span><span class="co">#&gt; 3002523                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002524                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002525                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002526                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt; 3002527                                            aminoglycoside antibiotic</span></span>
<span><span class="co">#&gt;                 Resistance Mechanism CARD Short Name</span></span>
<span><span class="co">#&gt; 3005099 antibiotic target alteration  Spyo_ErmA_MLSb</span></span>
<span><span class="co">#&gt; 3002523      antibiotic inactivation      AAC(2')-Ia</span></span>
<span><span class="co">#&gt; 3002524      antibiotic inactivation      AAC(2')-Ib</span></span>
<span><span class="co">#&gt; 3002525      antibiotic inactivation      AAC(2')-Ic</span></span>
<span><span class="co">#&gt; 3002526      antibiotic inactivation      AAC(2')-Id</span></span>
<span><span class="co">#&gt; 3002527      antibiotic inactivation      AAC(2')-Ie</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="参考文献">参考文献<a class="anchor" aria-label="anchor" href="#%E5%8F%82%E8%80%83%E6%96%87%E7%8C%AE"></a></h1>
<ol style="list-style-type: decimal"><li><p>Patil, K. R. &amp; Nielsen, J. Uncovering transcriptional regulation of metabolism by using metabolic network topology. Proc Natl Acad Sci U S A 102, 2685–2689 (2005).</p></li>
<li><p>L. Liu, R. Zhu, D. Wu, Misuse of reporter score in microbial enrichment analysis. iMeta. 2, e95 (2023).</p></li>
<li><p><a href="https://github.com/wangpeng407/ReporterScore" class="external-link uri">https://github.com/wangpeng407/ReporterScore</a></p></li>
</ol></div>


  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Chen Peng.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer></div>


<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({


    apiKey: '599cec31baffa4868cae4e79f180729b',
    indexName: 'docsearch',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script></body></html>

