---
title: "ReporterScore"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{ReporterScore}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

```{r setup}
library(ReporterScore)
```


# ReporterScore

Generalized Reporter Score-based Enrichment Analysis for Omics Data

## Citation

To cite ReporterScore in publications use:

C. Peng, Q. Chen, S. Tan, X. Shen, C. Jiang, Generalized Reporter Score-based Enrichment Analysis for Omics Data. bioRxiv [Preprint] (2023). https://doi.org/10.1101/2023.10.13.562235.

Please go to https://github.com/Asa12138/ReporterScore for more details.

## Install

```{r eval=FALSE}
if (!require("devtools")) install.packages("devtools")
devtools::install_github("Asa12138/pcutils")
devtools::install_github("Asa12138/ReporterScore")
library(ReporterScore)
```

## Usage

### 1. Inputdata (KO abundance table and metadata)

For transcriptomic, scRNA-seq, and related gene-based omics data of a specific species, a complete gene abundance table can be used. For metagenomic and metatranscriptomic data, which involve many different species, a KO abundance table can be used, generated using Blast, Diamond, or KEGG official mapper software to align the reads or contigs to the KEGG or the EggNOG database For metabolomic data, an annotated compound abundance table can be used, but the standardization of compound IDs (e.g., convert compound IDs to C numbers in the KEGG database) is required.

An example code tailored for a KO abundance table is as follows:

```{r}
data("KO_abundance_test")
head(KO_abundance[, 1:6])
```

And you should also offer a experimental metadata (rows are samples, columns are groups).

```{r}
head(metadata)
```

### 2. Pathway database

The `ReporterScore` package has built-in KEGG pathway, module, gene, compound, and GO databases and also allows customized databases, making it compatible with feature abundance tables from diverse omics data.

1. `ReporterScore` has built-in KEGG pathway-KO and module-KO databases (2023-08 version) for KO abundance table. You can use `load_KOlist()` to have a look and use `update_KO_file()` to update these databases (by KEGG API) as using latest database is very important.

2. `ReporterScore` has built-in KEGG pathway-compound and module-compound databases (2023-08 version) for compound abundance table. You can use `load_CPDlist()` to have a look and use `update_KO_file()` to update these databases (by KEGG API).

3. `ReporterScore` has built-in pathway-ko, pathway-gene, and pathway-compound databases of human (hsa) and mouse (mmu) for ko/gene/compound abundance table. You can use `custom_modulelist_from_org()` to have a look. Use `update_org_pathway()` to update these databases and download other organism databases (by KEGG API).

4. `ReporterScore` has built-in GO-gene database, You can use `load_GOlist()` to have a look and use `update_GO_file()` to update these databases (by KEGG API).

5. You can just customize your own pathway databases (gene set of interest) by using `custom_modulelist()`.

```{r eval=FALSE}
# 1. KEGG pathway-KO and module-KO databases
load_KOlist()
head(KOlist$pathway)

# 2. KEGG pathway-compound and module-compound databases
load_CPDlist()
head(CPDlist$pathway)

# 3. human (hsa) pathway-ko/gene/compound databases
hsa_pathway_gene <- custom_modulelist_from_org("hsa", feature = c("ko", "gene", "compound")[2])
head(hsa_pathway_gene)

# 4. GO-gene database
load_GOlist()
head(GOlist$BP)

# 5. customize your own pathway databases
?custom_modulelist()
```

### 3. One step enrichment

Use function `reporter_score` can get the reporter score result by one step.

there are some important arguments for analysis:

-   **mode**: "mixed" or "directed" (only for two groups differential analysis or multi-groups correlation analysis.), see details in `pvalue2zs`.
-   **method**: the type of test. Default is `wilcox.test`:
    -   `t.test` (parametric) and `wilcox.test` (non-parametric). Perform comparison between two groups of samples. If the grouping variable contains more than two levels, then a pairwise comparison is performed.
    -   `anova` (parametric) and `kruskal.test` (non-parametric). Perform one-way ANOVA test comparing multiple groups.
    -   "pearson", "kendall", or "spearman" (correlation), see `cor`.
-   **p.adjust.method**: p.adjust.method used for test result, see `p.adjust`.
-   **type**/**modulelist**: choose the pathway database, "pathway" or "module" for default database, or use a customized modulelist.
-   **feature**: one of "ko", "gene", "compound"

The first level will be set as the **control group**, you can change the factor level to change your comparison.

For example, we want to compare two groups 'WT-OE', and use the "directed" mode as we just want know which pathways are enriched or depleted in **OE group**:

```{r eval=TRUE,collapse=TRUE}
cat("Comparison: ", levels(factor(metadata$Group)))

# for real analysis, set the perm=9,999
reporter_score_res <- reporter_score(KO_abundance, "Group", metadata, mode = "directed", method = "wilcox.test", perm = 999)
```

The result is a "reporter_score" object:

|  elements    |  description                                      |
|--------------|---------------------------------------------------|
| `kodf`       | your input KO_abundance table                     |
| `ko_stat`    | ko statistics result contains p.value and z_score |
| `reporter_s` | the reporter score in each pathway                |
| `modulelist` | default KOlist or customized modulelist dataframe |
| `group`      | The comparison groups in your data                |
| `metadata`   | sample information dataframe contains group       |

### 4. Visualization

Plot the most significantly enriched pathways:

```{r fig.height=7,fig.width=12}
# View(reporter_score_res$reporter_s)
plot_report(reporter_score_res, rs_threshold = c(-2.5, 2.5))
```


Plot the most significantly enriched pathways (circle packing):

```{r fig.height=7}
# View(reporter_score_res$reporter_s)
plot_report_circle_packing(reporter_score_res, rs_threshold = c(-2.5, 2.5)) +
    coord_fixed()
```


When we focus on one pathway, e.g. "map00780":

```{r}
plot_KOs_in_pathway(reporter_score_res, map_id = "map00780")
```

Or show the distribution of Z-scores

```{r}
plot_KOs_distribution(reporter_score_res, map_id = "map00780")
```

Or display as a network:

```{r}
plot_KOs_network(reporter_score_res, map_id = c("map00780", "map00785", "map00900"), main = "", mark_module = TRUE)
```

And we also look at the KOs abundance in a pathway:

```{r fig.height=7,fig.width=8}
plot_KOs_box(reporter_score_res, map_id = "map00780", only_sig = TRUE)
```

Or display as a heatmap:

```{r }
plot_KOs_heatmap(reporter_score_res, map_id = "map00780", only_sig = TRUE, heatmap_param = list(cutree_rows = 2))
```

Or plot the KEGG pathway:

```{r,eval=FALSE}
plot_KEGG_map(reporter_score_res$ko_stat, map_id = "map00780", color_var = "Z_score")
```


### Example for multi-group or longitudinal

If our experimental design is more than two groups or longitudinal, we can choose multi-groups comparison (or correlation):

```{r collapse=TRUE}
cat("Comparison: ", levels(factor(metadata$Group2)))

reporter_score_res2 <- reporter_score(KO_abundance, "Group2", metadata,
    mode = "directed",
    method = "spearman", p.adjust.method1 = "none", perm = 999
)

plot_KOs_in_pathway(reporter_score_res2, map_id = "map02060") + scale_y_log10()
```

### Example for specified pattern

For example, groups “G1”, “G2”, and “G3” can be set as 1, 10, and 100 if an exponentially increasing trend is expected.

We use 1,5,1 to found pathways with the down-up-down pattern

```{r message=FALSE}
reporter_score_res3 <- reporter_score(KO_abundance, "Group2", metadata, mode = "directed", method = "pearson", pattern = c("G1" = 1, "G2" = 5, "G3" = 1), perm = 999)
plot_report(reporter_score_res3, rs_threshold = 3, show_ID = TRUE)
plot_KOs_in_pathway(reporter_score_res3, map_id = "map00860")
```

To explore potential patterns within the data, clustering methods, such as C-means clustering, can be used.

```{r message=FALSE}
rsa_cm_res <- RSA_by_cm(KO_abundance, "Group2", metadata, method = "pearson", k_num = 3, perm = 999)
# show the patterns
plot_c_means(rsa_cm_res, filter_membership = 0.7)

plot_report_bar(rsa_cm_res, rs_threshold = 2.5)
```

# Need helps?
If you have any questions/issues, please visit [github issue](https://github.com/Asa12138/ReporterScore/issues) tracker. 

# Session information

Here is the output of sessionInfo() on the system on which this document was compiled:

```{r}
sessionInfo()
```

# Reference

1.  Patil, K. R.
    & Nielsen, J. Uncovering transcriptional regulation of metabolism by using metabolic network topology.
    Proc Natl Acad Sci U S A 102, 2685--2689 (2005).

2.  L. Liu, R. Zhu, D. Wu, Misuse of reporter score in microbial enrichment analysis. iMeta. 2, e95 (2023).

3.  <https://github.com/wangpeng407/ReporterScore>
